---
title: "Analysis of RNAseq data in R and Bioconductor (part 4)<html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
subtitle: " <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html> Bioinformatics Resource Center - Rockefeller University"
author: 
  - "http://rockefelleruniversity.github.io/RU_RNAseq/"
  - "brc@rockefeller.edu"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```


```{r setup, include=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenomicAlignments)
library(DESeq2)
library(tximport)
library(org.Mm.eg.db)
library(goseq)
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# RNAseq (part 4)

---
"    
  )
  
}

```


## What we will cover

* Session 1: Alignment and counting_

* Session 2: Differential gene expression analysis

* Session 3: Visualizing results through PCA and clustering

* _Session 4: Gene Set Analysis_

* Session 5: Differential Transcript Utilization analysis

---
## The data

Over these sessions we will review data from Christina Leslie's lab at MSKCC on T-Reg cells. This can be found on the Encode portal. T-Reg data [here](https://www.encodeproject.org/experiments/ENCSR486LMB/) and activated T-Reg data [here](https://www.encodeproject.org/experiments/ENCSR726DNP/).

---
## The data

Several intermediate files have already been made for you to use. You can download the course content from GitHub [here](https://github.com/RockefellerUniversity/RU_RNAseq/archive/master.zip). Once downloaded you will need to unzip the folder and then navigate into the *r_course* directory of the download. This will mean the paths in the code are correct.

```{r, eval=F}
setwd("Path/to/Download/RU_RNAseq-master")


```


---
## The data

I have aligned all FQ to BAM and counted in genes and exons using Rsubread and summarizeOverlaps() and then analyzed this data for expression changes using DESeq2. You can find an excel file of differential expression analysis for this comparison in the **data** directory.

- CSV file of differential expression results for activated versus resting T-cells - **data/Group_Activated_minus_Resting.csv**.


---
## What we will cover

In our previous sessions we looked at how we can identify experimentally interesting genes. Either through finding genes with significant differential expression or performing clustering.

In this session we will explore a few ways we can evaluate any enrichment for functionally related genes. This helps us address common questions like:

**Do "cell cycle" genes change more between conditions then other genes?**

**Are genes related to "immune response" enriched in any clusters?**




---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Gene Sets

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Gene Sets

---
"    
  )
  
}

```

## What we test and how we test?

There are many options available to us for both functions/pathways/categories **(Gene set)** we will test and the methods we can use to test these gene sets for enrichment in our analysis **(Gene set Enrichment Analysis)**.

- **Gene set** - A named collection of genes.
- **Gene Set <u>e</u>nrichment Analysis - GSA** - A broad term for correlating a set of genes with a condition/phenotype.
- **Gene Set <u>E</u>nrichment Analysis - GSEA** - Broad Institute's term for correlating a set of genes with a condition/phenotype.


---
## Gene sets

Sources of well curated gene sets include [GO consortium](http://geneontology.org/) (gene's function, biological process and cellular localization), [REACTOME](http://www.reactome.org/) (Biological Pathways) and [MsigDB](http://software.broadinstitute.org/gsea/msigdb/) (Computationally and Experimentally derived).

You can also use a user-defined gene set i.e. genes of interest from a specific paper.

---
## Gene Ontology

The [Gene Ontology consortium](http://geneontology.org/) aims to provide a comprehensive resource of all the currently available knowledge regarding the functions of genes and gene products.

Functional categories of genes are broadly split into three main groups:

* **Molecular functions.** - Activity of a gene's protein product.
* **Biological processes.** - Role of gene's protein product.
* **Cellular components.** - Where in cell molecular function of protein product is performed.

---
## Gene Ontology

The three sub categories of gene ontology are arranged in nested, structured graph with gene sets at the top of graph representing more general terms and those at the bottom more specific terms.

![](imgs/goGraph.png)


---
## Reactome and KEGG

The [Reactome](http://www.reactome.org/) and [KEGG (Kyoto Encyclopedia of genes and genomes)](http://www.genome.jp/kegg/kegg2.html) contains information on genes' membership and roles in  molecular pathways.

These databases focus largely on metabolic and disease pathways, and allow us to investigate our genes in the context of not only functional roles but relative positions within pathways.


<div align="center">
<img src="imgs/map01100.png" alt="offset" height="300" width="600">
</div>


---
## MsigDB
.pull-left[
The [molecular signatures database (MsigDB)](http://software.broadinstitute.org/gsea/msigdb/) is available from the Broad institute and provides a set of curated gene sets derived from sources such as GO, pathway databases, motif scans and even other experimental sets.

MsigDB databases are widely used in Gene Set Enrichment Analysis and are available as plain text, in formats used with the popular Java based gene set enrichment software GSEA.
  ]

.pull-right[
<div align="center">
<img src="imgs/msigdb.png" alt="offset" height="500" width="350">
</div>
  ]


---
## Gene sets in Bioconductor

In R we can access information on these gene sets through database libraries (such as the Org.db.eg we have reviewed) such as **GO.db**, **reactome.db** or by making use of libraries which allows us to our gene sets from parse plain text formats, **GSEABase**. 

In our [Bioconductor sessions](https://rockefelleruniversity.github.io/Bioconductor_Introduction/presentations/singlepage/GenomicFeatures_In_Bioconductor.html#Gene_Annotation) we reviewed how we to get information on genes using the **Org.db** packages. We can access these dbs in similar ways. 

```{r,eval=T,echo=T, eval=F, echo=T, warning=FALSE,tidy=T}
library(GO.db)
library(reactome.db)
library(GSEABase)

```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# MSigDB and Gene Set Collections

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# MSigDB and gmt files

---
"    
  )
  
}
```

## GSEA and MSigDB

The GSEA software and MSigDB gene set collections can be found at the [Broad's site](http://software.broadinstitute.org/gsea/index.jsp).

We have access to Human gene sets containing information on:

* H	- hallmark gene sets
* C1	- positional gene sets
* C2	- curated gene sets
* C3	- motif gene sets
* C4	- computational gene sets
* C5	- GO gene sets
* C6	- oncogenic gene sets 
* C7	- immunologic gene sets

---
## Msigdbr

The MSigDB collection has been wrapped up in the [*msigdbr* CRAN package](https://igordot.github.io/msigdbr/). This package also contains computationally predicted homologs for a number of common species, so you can looks at these MSigDb groups in other organisms.

The msigdbr function is used to specify which organism and which categories you want. In return you get a type of data frame called a tibble. There is a lot of information in this data frame. 


```{r, msigdb}
library(msigdbr)

mm_H <- msigdbr(species = "Mus musculus", category = "H")
head(mm_H)
```


---
## Msigdbr and GO

If you are interested in GO terms we can grab them using the same approach. Here we get the GO terms but for Pig. 

```{r, wehiA }
library(msigdbr)

ss_C5 <- msigdbr(species = "pig", category = "C5")
head(ss_C5)
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Testing gene set enrichment

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Testing gene set enrichment

---
"    
  )
  
}
```


## Testing gene sets

We will review two different methods to identify functional groups associated with our condition of interest.

* The first method will test for any association of our gene set with our a group of interesting genes (differentially expressed genes/gene cluster) - Functional enrichment.
* The second method will test for any association of our gene set with high/low ranked genes (ranked by measure of differential expression) - GSEA.

---
## DE results from DESeq2

First lets read in our differential expression results from DESeq2 analysis of activated vs resting T-cells.

```{r, deres}
Activated_minus_Resting <- read.delim(file="data/Group_Activated_minus_Resting.csv",sep=",")
Activated_minus_Resting[1:3,]
```

---

## Background

All GSA methods will require us to filter to genes to those tested for differential expression. 

We will therefore filter to all genes which pass our independent filtering step from DESeq2. These will be our genes with no NA in the DESeq2 padj column.

```{r, deresFilter}
Activated_minus_Resting <- Activated_minus_Resting[!is.na(Activated_minus_Resting$padj),]
Activated_minus_Resting[1:3,]

```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Functional enrichment

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Functional enrichment

---
"    
  )
  
}
```

## Functional enrichment

When we run a functional enrichment analysis (a.ka. over representation analysis) we are testing if we observe a greater proportion of genes of interest in a gene set versus the proportion in the background genes. We do this with a Fisher exact test.

Example:

We have 100 genes of interest. 1/4 are in a gene set we care about. Is it enriched?

```{r, echo=F}

df <- data.frame( "Genes of Interest" = c(25,75),
                  "Gene Universe" = c(50,1950))
rownames(df) <- c("In Gene Set", "Out of Gene Set")

knitr::kable(df, caption = "Contingency Table for Fisher Test")

```

---
## ClusterProfiler

The **clusterProfiler** package provides multiple enrichment functions that work with curated gene sets (e.g. GO, KEGG) or custom gene sets. The most common way to use clusterProfiler is to run our over enrichment test using the one sided Fisher exact test.

clusterProfiler has both GSA and GSEA approaches, so is a nice one stop shop for everything. Plus it has some really nice and easy visualization options. Detailed information about all of the functionality within this package is available [here](http://yulab-smu.top/clusterProfiler-book/).

```{r, warning=F, message=F}
library(clusterProfiler)
```


---
## Running ClusterProfiler

First we need our gene information. For a simple functional enrichment of GO terms there is the *enrichGO()* function. You just provide this a vector of gene IDs you want to check, and the Org.db of the relevant organism. There are other specialized gene set functions i.e. *enrichKEGG()* for KEGG gene sets, or a generalized function *enrichr()* for any user provided gene sets.

```{r, warning=F, message=F}

sig_genes <- Activated_minus_Resting[Activated_minus_Resting$padj<0.05,1]
head(sig_genes)
```

```{r, warning=F, message=F}
sig_gene_enr <- enrichGO(sig_genes, OrgDb = org.Mm.eg.db)

```

---
## enrichResult object

The result is special object which contains the outcome of the functional enrichment. When you look at what is inside you get a preview of what test was done and what the results look like. 

```{r}
sig_gene_enr

```

---
## enrichResult object

You can also convert this to a dataframe and export.


```{r}
gsa_out <- as.data.frame(sig_gene_enr)
rio::export(gsa_out, "gsa_result.xlsx")

head(gsa_out)
```


---
## Visualizing the result

There are a couple of easy functions to visualize the top hits in your result. For example this dotplot which shows amount of overlap with geneset. These are also all built in ggplot2, so it is easy to modify parameters. For more info on ggplot2 check our course [here](https://rockefelleruniversity.github.io/Plotting_In_R/presentations/singlepage/ggplot2.html).

```{r, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
library(ggplot2)
dotplot(sig_gene_enr, showCategory = 6) + theme( axis.text.y = element_text(size = 7))

```

---
## Visualizing the result

Other useful visualizations include enrichment maps. These network plots show how the significant groups in the gene sets relate to each other. 

```{r, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
library(enrichplot)
sig_gene_enr <- pairwise_termsim(sig_gene_enr)
emapplot(sig_gene_enr, showCategory = 15, cex_label_category=0.6) + theme( text = element_text(size = 7))
```

---
## Beyond GO

There are several in-built functions that allow you to test against curated gene sets like GO terms (enrichGO) or KEGG (enrichKEGG). 

If we want to use our own gene sets or MSigDB we can provide them to the generic function *enricher()*.

First lets grab the curated gene sets (C2) from MSigDB using *msigdbr*. We only need the Gene Set Name and the Gene ID. 


```{r}

mm_c2 <- msigdbr(species = "Mus musculus", category = "C2")[,c("gs_name","entrez_gene")]
head(mm_c2)
```

---
## Beyond GO

We can then supply this data frame to *enricher()* as the TERM2GENE argument.

```{r, warning=F, message=F, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
sig_gene_enr <- enricher(sig_genes, TERM2GENE = mm_c2)
dotplot(sig_gene_enr)

```


---
## What universe?

*enricher()* also helps control your universe. The term universe refers to the background genes i.e. all the genes in your experiment. 

By default most of these tools will consider the background to be all genes in you gene sets. But a lot of these may not be in your experiment because they have been filtered or were not part of your annotation. 

To avoid inflating the significance it is good to make sure the universe is just the genes in your experiment. 

---
## What universe?

Lets update our universe to be the genes from our differential results. Remember this has been subset based on an NA value in the padj from DESeq2. 

It is always important to think not just about your genes of interest, but also what the background is. 

```{r, warning=F, message=F, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
sig_gene_enr <- enricher(sig_genes, TERM2GENE = mm_c2, universe = as.character(Activated_minus_Resting$ENTREZID))
dotplot(sig_gene_enr)



```


---
## Going further

clusterProfiler has a bunch of additional options for controlling your enrichment test. 

* pvalueCutoff - change the cutoff for what is significant. 
* pAdjustMethod - change multiple testing methods. 
* qvalueCutoff - change the cutoff for what is significant after multiple testing. 
* minGSSize/maxGSSize - exclude genes sets that are too big/large. 

---
## Extra Tools

We have focused on clusterProfiler as our package of choice for running these enrihcment analyses. For most use cases it is sufficient, but there are some other tools with specialist functions:

* [go.seq](https://bioconductor.org/packages/release/bioc/html/goseq.html) - This incorporates gene length into the null model to help mitigate any bias
* [topGO](https://bioconductor.org/packages/release/bioc/html/topGO.html) - If you are interested in GO terms and are only getting big parent terms i.e. "Cellular Process", topGO has alternative tests which weight the results to smaller child GO terms. 


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# GSEA

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# GSEA

---
"    
  )
  
}
```


## GSEA

Another popular method for testing gene set enrichment with differential expression analysis results is the Broad's GSEA method.

GSEA tests whether our gene set is correlated with the ranking of genes by our differential expression analysis metric.

![](imgs/GSEA2.png){width=30%}

---
## How it works

GSEA walks down the ranked gene list, increasing a score when a gene is in the gene set, decreasing it when it’s not. The resulting distribution is then tested versus a null distribution using a modified KS-test. The results is a significance score, an enrichment score (based on the size of the peak), and leading edge genes (based on all genes up from the peak).

![](imgs/GSEA3.png){width=50%}


---
## clusterProfiler inputs

We can run GSEA easily with clusterProfiler. First we will need to produce a ranked and named vector of gene scores. Here will rank by **stat** column to give sensible measure of differential expression. We could also use **log2FoldChange** column if we have modified log2 fold changes using **lfsShrink()** function.

```{r, myRNK}

forRNK <- Activated_minus_Resting$stat
names(forRNK) <- Activated_minus_Resting$ENTREZID
  
forRNK <- forRNK[order(forRNK, decreasing = T)]

forRNK[1:6]
```

---
## clusterProfiler inputs

**clusterProfiler** though which can just use a simple data frame. For this we will look at the MSigDb enrichment of C7 (immunological signature). clusterProfiler needs a 2 column data frame with the geneset names and gene IDs, just as we used for our over representation analysis we did earlier. 

```{r}

mm_c7 <- msigdbr(species = "Mus musculus", category = "C7")[,c("gs_name","entrez_gene")]
head(mm_c7)
```



---
## Running clusterProfiler
To run a GSEA there is a *GSEA()* function. We just provide our ranked list from differential gene expression analysis and the MSigDb Term2Gene dataframe.

```{r, warning=F, message=F}
sig_gene_enr <- GSEA(forRNK, TERM2GENE = mm_c7)#, eps=1e-100)
```


---
## Visualizing the result

We can easily visualize these GSEA results using the same methodology as GSAs to get dotplots and emaps. 

```{r, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
clusterProfiler::dotplot(sig_gene_enr, showCategory = 6) + theme( axis.text.y = element_text(size = 7))

```

---
## Visualizing the result

```{r, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
library(enrichplot)
sig_gene_enr <- pairwise_termsim(sig_gene_enr)
emapplot(sig_gene_enr, showCategory = 10, cex_label_category=0.6) + theme( text = element_text(size = 7))
```

---
## Visualizing the result

An additional visualization for GSEA is the Running Score plot. These are plotted for individual gene sets. here we are looking at the most signifcant group:


```{r, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
gseaplot(sig_gene_enr, geneSetID = 1, by = "runningScore", title = "GSE15330_HSC_VS_LYMPHOID_PRIMED_MULTIPOTENT_PROGENITOR_DN")


```

---
## fgsea

[**fgsea**](https://bioconductor.org/packages/release/bioc/html/fgsea.html) is an alternative tool to run GSEA. If you are running a lot of these tests or want a greater level of customization of the testing parameters it could be a better option. 

It doesn't have the same amount of plotting options, but it it very very fast and also gives lower level control.


---
## GSEA vs Fisher test

We have presented two options for testing for gene set enrichment. 

Fisher:
* Great for curated gene sets like clusters or filtered DGE results

GSEA: 
* Great for subtle coordinated changes and avoids using arbitrary cutoffs

If you're unsure, try both and review the results. 

---
## Time for an exercise


[Link_to_exercises](../../exercises/exercises/RNAseq_part4_exercise.html)

[Link_to_answers](../../exercises/answers/RNAseq_part4_answers.html)

