---
title: "Analysis of RNAseq data in R and Bioconductor (part 2)"
subtitle: " <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html> Bioinformatics Resource Center - Rockefeller University"
author: 
  - "http://rockefelleruniversity.github.io/RU_RNAseq/"
  - "brc@rockefeller.edu"
output: 
  xaringan::moon_reader:
    css: ["default.css", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
      titleSlideClass: [middle, inverse]
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)

# condaSalmon <- CondaSysReqs::install_CondaTools("salmon","salmon")
# pathToSalmon <- file.path(dirname(dirname(condaSalmon$pathToConda)),"envs",condaSalmon$environment,"bin","salmon")
```

```{r neQants,include=FALSE,eval=FALSE}

salmonExec <- paste0(pathToSalmon," quant")
fq <- "ENCFF070QMF.fastq.gz"
outDir <- "TReg_2_Quant"
salmonQuantCmd <- paste(salmonExec,
                        "-i",indexName,
                        "-o",outDir,
                        "-l A",
                        "-r",fq)
salmonQuantCmd
system(salmonQuantCmd, wait = TRUE)

salmonExec <- paste0(pathToSalmon," quant")
fq <- "ENCFF144YYI.fastq.gz"
outDir <- "TReg_act_1_Quant"
salmonQuantCmd <- paste(salmonExec,
                        "-i",indexName,
                        "-o",outDir,
                        "-l A",
                        "-r",fq)
salmonQuantCmd
system(salmonQuantCmd, wait = TRUE)

salmonExec <- paste0(pathToSalmon," quant")
fq <- "ENCFF042XBW.fastq.gz"
outDir <- "TReg_act_2_Quant"
salmonQuantCmd <- paste(salmonExec,
                        "-i",indexName,
                        "-o",outDir,
                        "-l A",
                        "-r",fq)
salmonQuantCmd
system(salmonQuantCmd, wait = TRUE)

salmonExec <- paste0(pathToSalmon," quant")
fq <- "ENCFF053CFZ.fastq.gz"
outDir <- "TReg_act_3_Quant"
salmonQuantCmd <- paste(salmonExec,
                        "-i",indexName,
                        "-o",outDir,
                        "-l A",
                        "-r",fq)
salmonQuantCmd
system(salmonQuantCmd, wait = TRUE)


```

```{r setup, include=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenomicAlignments)
library(DESeq2)
library(tximport)
library(org.Mm.eg.db)
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# RNAseq (part 2)

---
"    
  )
  
}

```


## What we will cover

* Session 1: Alignment and counting

* _Session 2: Differential gene expression analysis_

* Session 3: Visualizing results through PCA and clustering

* Session 4: Gene Set Analysis

* Session 5: Differential Transcript Utilization analysis

---
## The data

Over these sessions we will review data from Christina Leslie's lab at MSKCC on T-Reg cells. This can be found on the Encode portal. T-Reg data [here](https://www.encodeproject.org/experiments/ENCSR486LMB/) and activated T-Reg data [here](https://www.encodeproject.org/experiments/ENCSR726DNP/).

---
## The data

Several intermediate files have already been made for you to use. You can download the course content from GitHub [here](https://github.com/RockefellerUniversity/RU_RNAseq/archive/master.zip). Once downloaded you will need to unzip the folder and then navigate into the *r_course* directory of the download. This will mean the paths in the code are correct.

```{r, eval=F}
setwd("Path/to/Download/RU_RNAseq-master/r_course")


```

---
## The data

All Gene and Exon counts can be found as .RData obects in **data** directory

- Counts in genes can be found at - **data/GeneCounts.Rdata**
- Counts in disjoint exons can be found at - **data/ExonCounts.Rdata**

- Salmon transcript quantification output directories can be found under - **data/Salmon/**



---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Counting with multiple RNAseq datasets

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Counting with multiple RNAseq datasets

---
"    
  )
  
}

```

## Counts from SummarizeOverlaps

In our last session we counted our reads in genes using the summarize overlaps function to generate our **RangedSummarizedExperiment** object. We did this with a single BAM file.

To be able to do differntial gene expression analysis we need replicates. We can use similar approaches for handle multiple files. First I am using a **BamFileList** of all the T-Reg samples to easily organize my samples and save memory with the **yieldSize** parameter.

```{r eval=FALSE,echo=TRUE}
library(Rsamtools)
bamFilesToCount <- c("Sorted_Treg_1.bam","Sorted_Treg_2.bam",
                     "Sorted_Treg_act_1.bam","Sorted_Treg_act_2.bam",
                     "Sorted_Treg_act_3.bam")
names(bamFilesToCount) <- c("Sorted_Treg_1","Sorted_Treg_2",
                     "Sorted_Treg_act_1","Sorted_Treg_act_2",
                     "Sorted_Treg_act_3")
myBams <- BamFileList(bamFilesToCount,yieldSize = 10000)

```

---
## Counts from SummarizeOverlaps

We can count over this BamFileList in the same way we did for a single sample to produce our **RangedSummarizedExperiment** object of counts in genes across all samples.

```{r eval=FALSE,echo=TRUE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenomicAlignments)
geneExons <- exonsBy(TxDb.Mmusculus.UCSC.mm10.knownGene,by="gene")
geneCounts <- summarizeOverlaps(geneExons,myBams,
                                    ignore.strand = TRUE)
geneCounts
```

```{r gC1,eval=TRUE,echo=FALSE}
load("data/GeneCounts.RData")
geneCounts
```

---
## Speeding up SummarizeOverlaps

Bioconductor has a simple, unified system for parallelization through the [**BiocParallel** package.]("https://bioconductor.org/packages/release/bioc/html/BiocParallel.html") The BiocParallel package allows for differing parallel approaches including multiple cores on the same machine and distributed computing across large cluster environments (i.e a HPC).

Simply by installing and loading library, we are ready for parallelization.

```{r bp,eval=FALSE,echo=TRUE}
library(BiocParallel)
```

---
## Speeding up SummarizeOverlaps

We can control the parallelization in BiocParallel often choosing either a serial mode (no parallelization) with **SerialParam()** function or the number of cores to use with **MulticoreParam(workers=NUMBEROFCORES)**. 

We use the **register()** function to set the desired parallelization. Once registered you can then start counting using a parallelization. 

```{r bp1,eval=FALSE,echo=TRUE}
paramMulti <- MulticoreParam(workers=2)
paramSerial <- SerialParam()
register(paramSerial)
```

---
## Counts from SummarizeOverlaps

The counts object has already been made for you to save time. It is in the *data* directory

The counrs object is a **RangedSummarizedExperiment**, which contains our counts as a matrix accessible by the **assay()** function.

Each row is named by its Entrez gene ID.

```{r,eval=F}
load("data/GeneCounts.RData")

```
```{r gC2,eval=TRUE,echo=TRUE}
assay(geneCounts)[1:2,]
```

---
## Counts from SummarizeOverlaps

We can retrieve the GRangesList we counted by the **rowRanges()** accessor function.

Each GRangesList element contains exons used in counting, has Entrez gene ID and is in matching order as rows in our count table.


```{r gC3,eval=TRUE,echo=TRUE}
rowRanges(geneCounts)[1:2,]
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Differential gene expression analysis

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Differential gene expression analysis

---
"    
  )
  
}

```

## Differential expression with RNAseq

To perform differential expression between our sample groups we will need to perform some essential steps.

- Transform and/or normalize our counts.

- Calculate significance of differences between groups from normalized/transformed counts.


---
## Log transform or as counts

Two broadly different approaches to RNAseq analysis could be. 

- Log2 transform, normalize, use standard test for normal distribution to find differential expression.

- Normalize and use test suitable for count data to find differential expression.


---
## Normalization

As with microarrays, for the majority of normalization approaches we assume that the composition of RNA species across samples is similar and total RNA levels as the same.

To account for technical differences in library composition, many of these approached identify scaling factors for each sample which may be used to normalized observed/measured counts.

Simple measures such as reads per million (RPM) are scaled to total mapped reads. More sophisticated methods, such as used in the package **DESeq2** find scaling factors from the distribution of genes relative log expression versus their respective average across all samples.

.pull-left[
<div align="center">
<img src="imgs/beforeNorm.png" alt="offset" height="200" width="400">
</div>

  ]
.pull-right[
<div align="center">
<img src="imgs/afterNorm.png" alt="offset" height="200" width="400">
</div>

  ]

---
## Mean and variance relationship

Most approaches seek to model the relationship between mean expression and variance of expression in order to model and shrink variance of individual genes. This will assist us to detect differential expression with a small number of replicates.

![](imgs/meanVariance.png)


---
## Differential Expression Packages

Options for differential expression analysis in RNAseq data are available in R including DESeq2, EdgeR and Voom/Limma.

Both EdgeR and DESeq2 works with non-transformed count data to detect differential expression.

Voom transforms count data into log2 values with associated mean dependent weighting for use in Limma differential expression functions.

We will use the DESeq2 software to identify differential expression.

---
## DESeq2

To use DESeq2 we must first produce a dataframe of our Sample groups.

The DESeq2 package contains a workflow for assessing local changes in fragment/read abundance between replicated conditions. This workflow includes normalization, variance estimation, outlier removal/replacement as well as significance testing suited to high throughput sequencing data (i.e. integer counts).

To use DESeq2 we must first produce a dataframe of our Sample groups.

```{r de1,eval=TRUE,echo=TRUE}
metaData <- data.frame(Group=c("Naive","Naive","Act","Act","Act"),
                       row.names = colnames(geneCounts))
metaData
```


---
## Creating a DESeq2 object

We can use the **DESeqDataSetFromMatrix()** function to create a DESeq2 object.

We must provide our matrix of counts to countData parameter, our metadata data.frame to colData parameter and we include to an optional parameter of rowRanges the non-redundant peak set we can counted on.

Finally we provide the name of the column in our metadata data.frame within which we wish to test to the design parameter.


```{r de22,eval=TRUE,echo=TRUE, warning=F}
countMatrix <- assay(geneCounts)
countGRanges <- rowRanges(geneCounts)
dds <- DESeqDataSetFromMatrix(countMatrix,
                              colData = metaData,
                              design = ~Group,
                              rowRanges=countGRanges)
dds
```


---
## Creating a DESeq2 object

As an alternative input, we can first update our **RangedSummarizedExperiment** object to include our metadata using the **colData** accessor.

```{r de2,eval=TRUE,echo=TRUE}
colData(geneCounts)$Group <- metaData$Group
geneCounts
```

---
## Creating a DESeq2 object

The most recent version of the DESeq2 tools have a bug at the moment caused by a dependency: SummarizedExperiment. It is a little complicated, but you may see this error when creating DESeq2 objects, because R doesn't fully understand what they are. 

```
Error in validObject(.Object) : 
  invalid class “DESeqDataSet” object: superclass "ExpData" not defined in the environment of the object's class
```

There is a fix by running:

```{r}
setClassUnion("ExpData", c("matrix", "SummarizedExperiment"))
```



---
## Creating a DESeq2 object

Now we can make use of the  **DESeqDataSet()** function to build directly from our **RangedSummarizedExperiment** object.

We simply have to specify the **design** parameter to specify metadata column to test on.

```{r de3,eval=TRUE,echo=TRUE, warning=F}
dds <- DESeqDataSet(geneCounts,design = ~Group)
dds
```

---
## DESeq function

We can now run the DESeq2 workflow on our DESeq2 object using the **DESeq()** function. This function will normalize library sizes, estimate and shrink variance and test our data in a single step.

Our DESeq2 object is updated to include useful statistics such our normalized values and variance of signal within each gene.

```{r de4,eval=TRUE,echo=TRUE}
dds <- DESeq(dds)
```


---
## Normalization

In its first step the  **DEseq()** function normalizes our data by evaluating the median expression of genes across all samples to produce a per sample normalization factor (library size). 

We can retrieve normalized and unnormalized values from our DESeq2 object using the **counts()** function and specifying the **normalized** parameter as TRUE. Normalized counts are counts divided by the library scaling factor.

```{r de5,eval=TRUE,echo=TRUE}
normCounts  <- counts(dds, normalized=TRUE)
normCounts[1:2,]
```

---
## Variance Estimation

In the second step the **DEseq()** function estimates variances and importantly shrinks variance depending on the mean. This shrinking of variance allows us to detect significant changes with low replicate number.

We can review the Variance/mean relationship and shrinkage using **plotDispEsts()** function and our **DESeq2** object. We can see the <span style="color:blue">**adjusted dispersions in blue**</span> and <span style="color:black">**original dispersion for genes in black**</span>.
```{r de6,eval=TRUE,echo=TRUE,fig.width=6,fig.height=4}
plotDispEsts(dds)
```

---
## DESeq results

Finally we can extract our contrast of interest using the **results()** function.

We must specify the **contrast** parameter, listing the metadata column of interest and the two groups to compare.

The resulting **DESeqResults** can be sorted by pvalue to list most significant changes at top of table.


```{r de7,eval=TRUE,echo=TRUE}
myRes <-results(dds,contrast = c("Group","Act","Naive"))
myRes <- myRes[order(myRes$pvalue),]
myRes[1:3,]
```

---
## DESeqResults objects

The **DESeqResults** object contains important information from our differential expression testing.

- **baseMean** - Mean normalised values across all samples.
- **log2FoldChange** - Log2 of fold change between groups being compared.
- **pvalue** - Significance of change between groups.
- **padj** - Significance of change between groups corrected for multiple testing.

---
## DESeqResults objects

With our **DESeqResults** object we can get a very quick overview of differential expression changes using the **summary** function.

```{r drssaas,eval=FALSE,echo=TRUE}
summary(myRes)
```

```{r dr1,eval=TRUE,echo=FALSE}
DESeq2::summary(myRes)
```



---
## DESeqResults objects

We can review the relationship between fold-changes and expression levels with a MA-Plot by using the **plotMA()** function and our DESeqResults object. 

```{r drccd1a,eval=FALSE,echo=TRUE}
plotMA(myRes)
```

```{r dr1a,eval=TRUE,echo=FALSE,fig.height=4,fig.width=7}
DESeq2::plotMA(myRes)
```

---
## DESeqResults objects

We downweight genes with high fold change but low significance (due to low counts/high dispersion) by using the **lfcShrink()** function as we have the **results()** function.  This allows us to now use the log2FC as a measure of significance of change in our ranking for analysis and in programs such as **GSEA**.

```{r dr1b,eval=TRUE,echo=TRUE,fig.height=4,fig.width=7,message=FALSE,warning=FALSE}
myRes_lfc <- lfcShrink(dds, coef =  "Group_Naive_vs_Act")
DESeq2::plotMA(myRes_lfc)
```

```{r dr1c,eval=TRUE,echo=FALSE}
myRes <-results(dds,contrast = c("Group","Act","Naive"))
```

---
## DESeqResults objects

We can also convert our DESeqResults objects into a standard data frame using the **as.data.frame** function.

```{r dr2,eval=TRUE,echo=TRUE}
myResAsDF <- as.data.frame(myRes)
myResAsDF[1:2,]
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Significance and Multiple Testing

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Significance and Multiple Testing

---
"    
  )
  
}

```


## Multiple testing 
.pull-left[
In our analysis we are testing 1000s of genes and identifing significance of each gene separately at a set confidence interval.

If we are 95% confident a gene is differentially expressed when we look at all genes we can expect 5% to be false.

When looking across all genes we apply a multiple testing correction to account for this.
  ]
.pull-right[
<div align="center">
<img src="imgs/significant.png" alt="offset" height="600" width="400">
</div>

  ]

---


## Multiple testing 

We could apply a multiple testing to our p-values ourselves using either the Bonferroni or Benjamini-Hockberg correction.

-  Bonferonni = (p-value of gene)*(Total genes tested)
-  Benjamini-Hockberg = ((p-value of gene)*(Total genes tested))/(rank of p-value of gene)

---

## Multiple testing 

The total number of genes tested has a large effect on these multiple correction methods.

*Independent* filtering may be applied to reduce number of genes tested such as removing genes with low expression and/or variance.

**We can not simply filter to differentially expressed genes and reapply correction. This is not independent!**


---
## NA in padj

By default DEseq2 filters out low expressed genes to assist in multiple testing correction. This results in **NA** values in **padj** column for low expressed genes.

```{r dr22a,eval=TRUE,echo=TRUE}
table(is.na(myResAsDF$padj))
```

---
## NA in padj

We can add a column of our own adjusted p-values for all genes with a pvalue using the **p.adjust()** function. 

```{r dr22b,eval=TRUE,echo=TRUE}
myResAsDF$newPadj <- p.adjust(myResAsDF$pvalue)
myResAsDF[1:3,]
```

---
## NA in padj

Typcially genes with **NA** padj values should be filtered from the table for later evaluation and functional testing.

```{r dr22,eval=TRUE,echo=TRUE}
myResAsDF <- myResAsDF[!is.na(myResAsDF$padj),]
myResAsDF <- myResAsDF[order(myResAsDF$pvalue), ]
myResAsDF[1:3,]
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# DEseq2 and Salmon

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# DEseq2 and Salmon

---
"    
  )
  
}

```


## DESeq2 from Salmon

We saw at the end of last session how we can gain transcript quantification from FASTQ using Salmon.

Salmon offers a fast alternative to alignment and counting for transcript expression estimation so we would be keen to make use of this in our differential expression analysis.

To make use of Salmon transcript quantifications for differential gene expression changes we must however summarize our transcript expression estimates to gene expression estimates.

---
## tximport 

The tximport package, written by same author as DESeq2, offers functions to import data from a wide range of transcript counting and quantification software and summarizes transcript expression to genes.

We can first load the Bioconductor [**tximport** package.](https://bioconductor.org/packages/release/bioc/html/tximport.html)

```{r tx1,eval=TRUE,echo=TRUE}
library(tximport)
```

---
## tximport 

In order to summarize our transcripts to genes we will need to provide a data.frame of transcripts to their respective gene names.

First we read in a single Salmon quantification report to get a table containing all transcript names.

```{r tx2,eval=TRUE,echo=TRUE}
temp <- read.delim("data/Salmon/TReg_2_Quant/quant.sf")

temp[1:3,]
```

---
## tximport 

We can now use the **select()** function from AnnotationDbi to retrieve a transcript names (TXNAME) to gene ids (GENEID) map from our TxDb.Mmusculus.UCSC.mm10.knownGene object. 

Remember that available columns and keys can be found using the **columns()** and **keytype** functions respectively.

```{r tx3a,eval=TRUE,echo=TRUE,warning=FALSE,message=FALSE}

Tx2Gene <- AnnotationDbi::select(TxDb.Mmusculus.UCSC.mm10.knownGene,
                  keys = as.vector(temp[,1]),
                  keytype = "TXNAME",
                  columns = c("GENEID","TXNAME"))
Tx2Gene <- Tx2Gene[!is.na(Tx2Gene$GENEID),]
Tx2Gene[1:10,]
```


---
## tximport 

Now we can use the **tximport** function to import and summarize our Salmon quantification files to gene level expression estimates.

We must provide the paths to Salmon *quant.sf* files, the type of file to import (here "salmon") to the **type** argument and our data.frame of transcript to gene mapping using the **tx2gene** argument.

```{r tx4,eval=TRUE,echo=TRUE}
salmonQ <- dir("data/Salmon/",recursive = T,
               pattern = "quant.sf",full.names = T)
salmonCounts <- tximport(salmonQ,
                         type="salmon",
                         tx2gene = Tx2Gene)
```

---
## tximport 

The result from tximport is a list containing our summarized gene expression estimates

**Abundance** - Number of transcripts per million transcripts (TPM).  
**Counts** - Estimated counts.


```{r tx5,eval=TRUE,echo=TRUE}
salmonCounts$abundance[1:2,]
salmonCounts$counts[1:2,]
```

---
## DESeq2 from Tximport 

We can now use the **DESeqDataSetFromTximport()** to build our DESeq2 object from salmon counts. We must specify the metadata to **colData** parameter and specify the column of interest in **design** parameter as we have done for **DESeqDataSetFromMatrix()**.

```{r tx6,eval=TRUE,echo=TRUE, warning=F}
ddsSalmon <- DESeqDataSetFromTximport(salmonCounts,
                                      colData = metaData,
                                      design = ~Group)
```


---
## DESeq2 from Tximport 

We can then proceed as we did for previous analysis using summarizeOverlaps counts.

```{r tx7,eval=TRUE,echo=TRUE,message=FALSE,warning=FALSE}
ddsSalmon <- DESeq(ddsSalmon)
myResS <-results(ddsSalmon,contrast = c("Group","Act","Naive"))
myResS <- myResS[order(myResS$pvalue),]
myResS[1:3,]
```


---
## Salmon and Rsubread

We can now compare the differential following Salmon quantification and Rsubread/summarizeOverlaps. Here we are showing the overlaps between genes that have a padj <0.05 and also a fold change greater then 2.   

```{r dr1ss,eval=TRUE,echo=FALSE}
swde <- merge(as.data.frame(myRes),as.data.frame(myResS),by=0,all=FALSE)

toCompare <- cbind(abs(swde$log2FoldChange.x) > 1 & swde$padj.x < 0.05 & !is.na(swde$padj.x) & !is.na(swde$padj.y), abs(swde$log2FoldChange.y) > 1 & swde$padj.y < 0.05 & !is.na(swde$padj.y) & !is.na(swde$padj.x))
colnames(toCompare) <- c("summarizeOverlaps","Salmon")
limma::vennDiagram(toCompare)
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Adding Annotation

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Adding Annotation

---
"    
  )
  
}

```

## Annotation of results table

By whatever method we have chosen to analyze differential expression we will want to add some sensible gene names or symbols to our data frame of results.

We can use the **org.db** packages to retrieve Gene Symbols for our Entrez IDs using the **select** function. Here we use the **org.db** package for mouse, **org.db**

```{r anno,eval=TRUE,echo=TRUE,message=FALSE,warning=FALSE,tidy=FALSE}
library(org.Mm.eg.db)
eToSym <- AnnotationDbi::select(org.Mm.eg.db,
                 keys = rownames(myResAsDF),
                 keytype = "ENTREZID",
                 columns="SYMBOL")
eToSym[1:10,]
```


---
## Annotation of results table

Now we can merge the Entrez ID to Symbol table into our table of differential expression results.

```{r anno2,eval=TRUE,echo=TRUE,tidy=FALSE}
annotatedRes <- merge(eToSym,myResAsDF,
                      by.x=1,
                      by.y=0,
                      all.x=FALSE,
                      all.y=TRUE)
annotatedRes <- annotatedRes[order(annotatedRes$pvalue),]
annotatedRes[1:3,]
```

---
## Time for an exercise


[Link_to_exercises](../../exercises/exercises/RNAseq_part2_exercise.html)

[Link_to_answers](../../exercises/answers/RNAseq_part2_answers.html)

