---
title: "Analysis of RNAseq data in R and Bioconductor (part 4)<html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
subtitle: " <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html> Bioinformatics Resource Center - Rockefeller University"
author: 
  - "http://rockefelleruniversity.github.io/RU_RNAseq/"
  - "brc@rockefeller.edu"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```


```{r setup, include=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenomicAlignments)
library(DESeq2)
library(tximport)
library(org.Mm.eg.db)
library(goseq)
```


## What we will cover

* Session 1: Alignment and counting_

* Session 2: Differential gene expression analysis

* Session 3: Visualizing results through PCA and clustering

* _Session 4: Gene Set Analysis_

* Session 5: Differential Transcript Utilization analysis

---
## The data

Over these sessions we will review data from Christina Leslie's lab at MSKCC on T-Reg cells. This can be found on the Encode portal. T-Reg data [here](https://www.encodeproject.org/experiments/ENCSR486LMB/) and activated T-Reg data [here](https://www.encodeproject.org/experiments/ENCSR726DNP/).

---
## The data

Several intermediate files have already been made for you to use. You can download the course content from GitHub [here](https://github.com/RockefellerUniversity/RU_RNAseq/archive/master.zip). Once downloaded you will need to unzip the folder and then navigate into the *r_course* directory of the download. This will mean the paths in the code are correct.

```{r, eval=F}
setwd("Path/to/Download/RU_RNAseq-master")


```


---
## The data

I have aligned all FQ to BAM and counted in genes and exons using Rsubread and summarizeOverlaps() and then analyzed this data for expression changes using DESeq2. You can find an excel file of differential expression analysis for this comparison in the **data** directory.

- CSV file of differential expression results for activated versus resting T-cells - **data/Group_Activated_minus_Resting.csv**.


---
## What we will cover

In our previous sessions we looked at how we can identify experimentally interesting genes. Either through finding genes with significant differential expression or performing clustering.

In this session we will explore a few ways we can evaluate any enrichment for functionally related genes. This helps us address common questions like:

**Do "cell cycle" genes change more between conditions then other genes?**

**Are genes related to "immune response" enriched in any clusters?**




---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Gene Sets

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Gene Sets

---
"    
  )
  
}

```

## What we test and how we test?

There are many options available to us for both functions/pathways/categories **(Gene set)** we will test and the methods we can use to test these gene sets for enrichment in our analysis **(Gene set Enrichment Analysis)**.

- **Gene set** - A named collection of genes.
- **Gene Set <u>e</u>nrichment Analysis - GSA** - A broad term for correlating a set of genes with a condition/phenotype.
- **Gene Set <u>E</u>nrichment Analysis - GSEA** - Broad Institute's term for correlating a set of genes with a condition/phenotype.


---
## Gene sets

Sources of well curated gene sets include [GO consortium](http://geneontology.org/) (gene's function, biological process and cellular localization), [REACTOME](http://www.reactome.org/) (Biological Pathways) and [MsigDB](http://software.broadinstitute.org/gsea/msigdb/) (Computationally and Experimentally derived).

You can also use a user-defined gene set i.e. genes of interest from a specific paper.

---
## Gene Ontology

The [Gene Ontology consortium](http://geneontology.org/) aims to provide a comprehensive resource of all the currently available knowledge regarding the functions of genes and gene products.

Functional categories of genes are broadly split into three main groups:

* **Molecular functions.** - Activity of a gene's protein product.
* **Biological processes.** - Role of gene's protein product.
* **Cellular components.** - Where in cell molecular function of protein product is performed.

---
## Gene Ontology

The three sub categories of gene ontology are arranged in nested, structured graph with gene sets at the top of graph representing more general terms and those at the bottom more specific terms.

![](imgs/goGraph.png)


---
## Reactome and KEGG

The [Reactome](http://www.reactome.org/) and [KEGG (Kyoto Encyclopedia of genes and genomes)](http://www.genome.jp/kegg/kegg2.html) contains information on genes' membership and roles in  molecular pathways.

These databases focus largely on metabolic and disease pathways, and allow us to investigate our genes in the context of not only functional roles but relative positions within pathways.


<div align="center">
<img src="imgs/map01100.png" alt="offset" height="300" width="600">
</div>


---
## MsigDB
.pull-left[
The [molecular signatures database (MsigDB)](http://software.broadinstitute.org/gsea/msigdb/) is available from the Broad institute and provides a set of curated gene sets derived from sources such as GO, pathway databases, motif scans and even other experimental sets.

MsigDB databases are widely used in Gene Set Enrichment Analysis and are available as plain text, in formats used with the popular Java based gene set enrichment software GSEA.
]
.pull-right[
<div align="center">
<img src="imgs/msigdb.png" alt="offset" height="500" width="350">
</div>
]


---
## Gene sets in Bioconductor

In R we can access information on these gene sets through database libraries (such as the Org.db.eg we have reviewed) such as **GO.db**, **reactome.db** or by making use of libraries which allows us to our gene sets from parse plain text formats, **GSEABase**. 

In our [Bioconductor sessions](https://rockefelleruniversity.github.io/Bioconductor_Introduction/presentations/singlepage/GenomicFeatures_In_Bioconductor.html#Gene_Annotation) we reviewed how we to get information on genes using the **Org.db** packages. We can access these dbs in similar ways. 

```{r,eval=T,echo=T, eval=F, echo=T, warning=FALSE,tidy=T}
library(GO.db)
library(reactome.db)
library(GSEABase)

```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# MSigDB and Gene Set Collections

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# MSigDB and gmt files

---
"    
  )
  
}
```

## GSEA and MSigDB

The GSEA software and MSigDB gene set collections can be found at the [Broad's site](http://software.broadinstitute.org/gsea/index.jsp).

We have access to Human gene sets containing information on:

* H	- hallmark gene sets
* C1	- positional gene sets
* C2	- curated gene sets
* C3	- motif gene sets
* C4	- computational gene sets
* C5	- GO gene sets
* C6	- oncogenic gene sets 
* C7	- immunologic gene sets

---
## MSigDB and gmt format

The **gmt** (gene matrix transposed) is a GSEA specific format which is easy to edit in spreadsheet programs.

It contains tab separated space with lines containing gene sets of differing sizes.

---
## GSEAbase

We can read gene sets stored in the gmt format into R using the GSEABase package. The GSEABase function **getGmt** allows us to import gmt formatted gene set collections into a **GeneSetCollection** object. In this case we are reading in the Hallmarks gene set with gene symbols which we downloaded from the [Broad website.](http://software.broadinstitute.org/gsea/index.jsp)

```{r ,gseabase, warning=F, message=F}
library(GSEABase)
hallMarks <- getGmt(con="data/h.all.v7.1.symbols.gmt")
hallMarks
```


---
## GeneSetCollection

The GeneSetCollection object can be subset as a list. We can access the names of gene sets using the **names()** accessor

```{r, GeneSetCollection}
hallMarks[[1]]
```

```{r}
names(hallMarks)
```

---
## GeneSetCollection

The geneIDs() function returns a list named by gene set with each named element containing the genes (their IDS) contained within that set.

```{r, geneIDs}
geneIds(hallMarks)[1:3]
```

---
## Msigdbr

If you do not need to work with *.gmt* files, the MSigDB collection has been wrapped up in the [*msigdbr* CRAN package](https://igordot.github.io/msigdbr/). This package also contains computationally predicted homologs for a number of common species, so you can looks at these MSigDb groups in other organisms.

The msigdbr function is used to specify which organism and which categories you want. In return you get a simple dataframe. 


```{r, wehiA }
library(msigdbr)

mm_H <- msigdbr(species = "Mus musculus", category = "H")
head(mm_H)
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Testing gene set enrichment

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Testing gene set enrichment

---
"    
  )
  
}
```


## Testing gene sets

We will review two different methods to identify functional groups associated with our condition of interest.

* The first method will test for any association of our gene set with our a group of interesting genes (differentially expressed genes) - functional enrichment.
* The second method will test for any association of our gene set with the ranking of all our genes (ranked by measure of differential expression) - GSEA.

---

## DE results from DESeq2

First lets read in our differential expression results from DESeq2 analysis of activated vs resting T-cells.

```{r, deres}
Activated_minus_Resting <- read.delim(file="data/Group_Activated_minus_Resting.csv",sep=",")
Activated_minus_Resting[1:3,]
```

---

## Background

All GSA methods will require us to filter to genes to those tested for differential expression. 

We will therefore filter to all genes which pass our independent filtering step from DESeq2. These will be our genes with no NA in the DESeq2 padj column.

```{r, deresFilter}
Activated_minus_Resting <- Activated_minus_Resting[!is.na(Activated_minus_Resting$padj),]
Activated_minus_Resting[1:3,]

```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Functional enrichment

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Functional enrichment

---
"    
  )
  
}
```

## goseq package

We can use the goseq package to identify any functional enrichment in a set of interesting genes (i.e. differentially expressed). The goseq package requires a named vector of 1s or 0s indicating whether a gene is upregulated/downregulated. 

Here we define a set of genes significantly upregulated in Activated vs Naive T-reg cells (padj <0 05 and log2FC > 0).

```{r func,eval=TRUE,echo=TRUE,cache=TRUE,dependson="anno2"}
UpInAct <- Activated_minus_Resting$padj < 0.05 & 
             Activated_minus_Resting$log2FoldChange > 0
UpInAct <- as.integer(UpInAct)
names(UpInAct) <- Activated_minus_Resting$ENTREZID
UpInAct[1:4]
table(UpInAct)
```



---
## Gene ontology and functional testing
We can see which genomes are supported using the **supportedGenomes()** function.

```{r,include=FALSE}
library(goseq)
```

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
supGenomes <- supportedGenomes()
supGenomes[1:2,]

```

---
## Gene length bias

The **goseq** package is specifically designed to account for potential length biases in differential expression analysis and so can remove any artefactual enrichment for long and short genes.

The **nullp** function assess any biases by evaluating the relationship between differential expression and length. 

```{r func1,eval=TRUE,echo=TRUE,cache=TRUE,dependson="func", warning=F, message=F, fig.height=3.75,fig.width=3.75}
library(goseq)
pwf <- nullp(UpInAct, "mm10", "knownGene", plot.fit = TRUE)
```

<div align="center">
<img src="imgs/fir.png" alt="igv" height="200" width="500">
</div>

---
## Functional Enrichment analysis

Having corrected for any bias we can now use the **goseq** function to for enrichment of GO or KEGG terms.

We must specify the genome build and ID we are using (here matching our TxDb.UCSC.**mm10**.**knownGene**.db) and the categories we wish to test (GO:BP, GO:MF, GO:CC, KEGG). 

Here we identify the top enriched functional terms in GO biological processes.
```{r funca,eval=TRUE,echo=FALSE,cache=TRUE,include=FALSE}
load(file = "data/fit.RData")
```

```{r func2,eval=TRUE,echo=TRUE,cache=TRUE,dependson="func1",warning=FALSE,message=FALSE}
GO_UpInAct <- goseq(pwf,"mm10","knownGene",
                       test.cats=c("GO:BP"))
GO_UpInAct[1:3,]
```

---
## Retrieving function groups

We can retrieve the genes in the **immune response (GO:0006955)** functional group using the **select** and the **org.Mm.eg.db** object.


```{r func3,eval=TRUE,echo=TRUE,cache=TRUE,dependson="funca",warning=FALSE,message=FALSE}
library(org.Mm.eg.db)
ImmuneResponseGenes <- AnnotationDbi::select(org.Mm.eg.db, keytype = "GOALL",
                              keys = "GO:0006955", columns = "ENTREZID")
ImmuneResponseGenes

```

---
## Retrieving function groups

We can then filter our DE table, with this GO set of immune response GENES and write to file.


```{r func4,eval=TRUE,echo=TRUE,cache=TRUE,dependson="func3",warning=FALSE,message=FALSE}
IRG_Entrez <- unique(ImmuneResponseGenes$ENTREZID)
IRG_Res <-  Activated_minus_Resting[Activated_minus_Resting$ENTREZID %in% IRG_Entrez,]
write.table(IRG_Res,
            file="data/ImmuneResponseGeneTable.csv",sep=",",
            row.names = FALSE)
IRG_Res[1:3,]
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# ClusterProfiler

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# ClusterProfiler

---
"    
  )
  
}
```


## ClusterProfiler

The **clusterProfiler** package provides multiple enrichment functions that work with curated gene sets (e.g. GO, KEGG) or custom gene sets. It can perform both GSA and GSEA approaches, so is a nice one stop shop for everything. Plus it has some really nice and easy visualization options. Detailed information about all of the functionality within this package is available [here](http://yulab-smu.top/clusterProfiler-book/).


```{r, warning=F, message=F}
library(clusterProfiler)
```


---
## Running ClusterProfiler

First we need our gene information. For a simple functional enrichment of GO terms there is the *enrichGO()* function. You just provide this a vector of gene IDs you want to check, and the Org.db of the relevant organism. There are other specialized gene set functions i.e. *enrichKEGG()* for KEGG gene sets, or a generalized function *enrichr()* for any user provided gene sets.

```{r, warning=F, message=F}

sig_genes <- Activated_minus_Resting[Activated_minus_Resting$padj<0.05,1]
head(sig_genes)
```

```{r, warning=F, message=F}
sig_gene_enr <- enrichGO(sig_genes, OrgDb = org.Mm.eg.db)

```

---
## enrichResult object

The result is special object which contains the outcome of the functional enrichment. When you look at what is inside you get a preview of what test was done and what the results look like. You can also convert this to a dataframe and export.

```{r}
sig_gene_enr

```

---
## Visualizing the result

There are a couple of easy functions to visualize the top hits in your result. For example this dotplot which shows amount of overlap with geneset. These are also all built in ggplot2, so it is easy to modify parameters. For more info on ggplot2 check our course [here](https://rockefelleruniversity.github.io/Plotting_In_R/presentations/singlepage/ggplot2.html).

```{r, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
library(ggplot2)
clusterProfiler::dotplot(sig_gene_enr, showCategory = 6) + theme( axis.text.y = element_text(size = 7))

```
---
## Visualizing the result

Other useful visualizations include enrichment maps. These network plots show how the signifcant groups in the gene sets relate to each other. 

```{r}
library(enrichplot)
sig_gene_enr <- pairwise_termsim(sig_gene_enr)
emapplot(sig_gene_enr, showCategory = 15, cex_label_category=0.6) + theme( text = element_text(size = 7))
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# GSEA

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# GSEA

---
"    
  )
  
}
```


## GSEA

Another popular method for differential expression analysis is the Broad's GSEA method.

GSEA tests whether our gene set is correlated with the ranking of genes by our differential expression analysis metric using a modified KS-test.

![](imgs/GSEA.jpg)

---
## clusterProfiler inputs

First we will need to produce a ranked and named vector of gene scores. Here will rank by **stat** column to give sensible measure of differential expression. We could also use **log2FoldChange** column if we have modified log2 fold changes using **lfsShrink()** function.

```{r, myRNK}

forRNK <- Activated_minus_Resting$stat
names(forRNK) <- Activated_minus_Resting$ENTREZID
  
forRNK <- forRNK[order(forRNK, decreasing = T)]

forRNK[1:6]
```

---
## clusterProfiler inputs

Often GSEA programs will use gmt files as this is the standard Broad format such as the R package **fgsea**. We will be using **clusterProfiler** though which can just use a simple data frame.
For this we will look at the MSigDb enrichment of C7 (immunological signature). clusterProfiler needs a 2 column dataframe with the geneset names and gene IDs. 

```{r}

mm_c7 <- msigdbr(species = "Mus musculus", category = "C7")[,c("gs_name","entrez_gene")]
head(mm_c7)
```



---
## Running clusterProfiler
To run a GSEA there is a *GSEA()* function. We just provide our ranked list from differential gene expression analysis and the MSigDb Term2Gene dataframe.

```{r, warning=F, message=F}
sig_gene_enr <- GSEA(forRNK, TERM2GENE = mm_c7, eps=1e-100)
```


---
## Visualizing the result

We can easily visualize these GSEA results using the same methodology as GSAs to get dotplots and emaps. 

```{r, fig.width=10, fig.height=4, message=FALSE, warning=FALSE}
clusterProfiler::dotplot(sig_gene_enr, showCategory = 6) + theme( axis.text.y = element_text(size = 7))

```

---
## Visualizing the result

```{r}
library(enrichplot)
sig_gene_enr <- pairwise_termsim(sig_gene_enr)
emapplot(sig_gene_enr, showCategory = 10, cex_label_category=0.6) + theme( text = element_text(size = 7))
```

---
## Visualizing the result

An additional visualization for GSEA is the Running Score plot. These are plotted for individual gene sets. here we are looking at the most signifcant group:


```{r}
gseaplot(sig_gene_enr, geneSetID = 1, by = "runningScore", title = "GSE15330_HSC_VS_LYMPHOID_PRIMED_MULTIPOTENT_PROGENITOR_DN")


```

## Exporting the result

You can export the result to an excel spreadsheet very easily. You just have to coerce the result object to a data frame. 

```{r}

write.csv(as.data.frame(sig_gene_enr), "cluster_profiler_GSEA_result.csv")

```


---
## Time for an exercise


[Link_to_exercises](../../exercises/exercises/RNAseq_part4_exercise.html)

[Link_to_answers](../../exercises/answers/RNAseq_part4_answers.html)

