<!DOCTYPE html>
<html>
  <head>
    <title></title>
    <meta charset="utf-8">
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Analysis of RNA-seq data in R and Bioconductor
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### <a href="http://rockefelleruniversity.github.io/RU_RNAseq/" class="uri">http://rockefelleruniversity.github.io/RU_RNAseq/</a>

---





                                 
## The data.

In this session we will be reviewing data from Christina Leslie's lab at MSKCC on T-Reg cells and activated T-Reg cells. T-Reg data can be found on the Encode portal [here](https://www.encodeproject.org/experiments/ENCSR486LMB/) and activated T-Reg data [here](https://www.encodeproject.org/experiments/ENCSR726DNP/)

I have aligned all FQ to BAM and counted in genes and exons using Rsubread and summariseOverlaps().

I have also quantified our transcripts from FQ files using our transcript fasta from last session and Salmon.

---
## The data.

All Gene and Exon counts can be found as .RData obects in **Data** directory

- Counts in genes can be found at - **Data/GeneCounts.RData**
- Counts in disjoint exons can be found at - **Data/ExonCounts.RData**

- Salmon transcript quantification output directories can be found under - **Data/Salmon/**

---
## What we will cover.

In our first session we will looked at two separate ways we can gain gene expression estimates from raw sequencing data as fastQ files.

In our last session we imported our expression estimates to identify differences in the expression of genes between two conditions. 

In this session we will look at how we can identify differences in the expression of genes between multiple conditions and how we can identify changes in Exon usage to infer changes in transcript usage.

---
## Counts from SummariseOverlaps

In our first session we counted our reads in genes using the summarise overlaps function to generate our **RangedSummarizedExperiment** object.

I have counted using a **BamFileList** of all T-Reg samples to easily organise my samples and save memory with the **yieldSize** parameter.


```r
library(Rsamtools)
bamFilesToCount &lt;- c("Sorted_Treg_1.bam","Sorted_Treg_2.bam",
                     "Sorted_Treg_act_1.bam","Sorted_Treg_act_2.bam",
                     "Sorted_Treg_act_3.bam")
myBams &lt;- BamFileList(bamFilesToCount,yieldSize = 10000)
```

---
## Counts from SummariseOverlaps

For differential exon usage we must use a non-overlapping,disjoint set of exons to count on.

We can produce a non-overlapping set of exons by using the **disjointExons()** function. This is equivalent of us having to create GFF file and using the python script wrapped up in DEXseq package (*dexseq_prepare_annotation.py*).


```r
library(GenomicFeatures)
nonOverlappingExons &lt;- disjointExons(TxDb.Mmusculus.UCSC.mm10.knownGene)
names(nonOverlappingExons) &lt;- paste(mcols(nonOverlappingExons)$gene_id,
                                    mcols(nonOverlappingExons)$exonic_part,
                                    sep="_")
nonOverlappingExons[1:2,]
```

```
## GRanges object with 2 ranges and 3 metadata columns:
##               seqnames               ranges strand |         gene_id
##                  &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;
##   100009600_1     chr9 [21062393, 21062717]      - |       100009600
##   100009600_2     chr9 [21062894, 21062987]      - |       100009600
##                             tx_name exonic_part
##                     &lt;CharacterList&gt;   &lt;integer&gt;
##   100009600_1 uc009veu.1,uc033jjg.1           1
##   100009600_2 uc009veu.1,uc033jjg.1           2
##   -------
##   seqinfo: 66 sequences (1 circular) from mm10 genome
```

---
## Disjoint Exons

![](imgs/collapsedE1.png)

---
## Disjoint Exons

![](imgs/collapsedE2.png)


---
## Counts from SummariseOverlaps

Now we have our disjoint exons to count on we can use the **summarizeOverlaps().**



```r
tregExonCounts &lt;- summarizeOverlaps(nonOverlappingExons,
                                    myBam,
                                    ignore.strand = TRUE,
                                    inter.feature=FALSE)
tregExonCounts
```


```
## class: RangedSummarizedExperiment 
## dim: 241668 5 
## metadata(0):
## assays(1): counts
## rownames(241668): 100009600_1 100009600_2 ... 99982_19 99982_20
## rowData names(3): gene_id tx_name exonic_part
## colnames(5): Sorted_T_reg_act_1.bam Sorted_T_reg_1.bam
##   Sorted_T_reg_act_2.bam Sorted_T_reg_2.bam Sorted_T_reg_act_3.bam
## colData names(0):
```

---
## Counts from SummariseOverlaps

Our **RangedSummarizedExperiment** contains our counts as a matrix accessible by the **assay()** function.

Each row is named by our Exon ID, geneID + exon number.



```r
assay(tregExonCounts)[1:2,]
```

```
##             Sorted_T_reg_act_1.bam Sorted_T_reg_1.bam
## 100009600_1                      0                  1
## 100009600_2                      2                  0
##             Sorted_T_reg_act_2.bam Sorted_T_reg_2.bam
## 100009600_1                      2                  1
## 100009600_2                      0                  1
##             Sorted_T_reg_act_3.bam
## 100009600_1                      4
## 100009600_2                      0
```
---
## Counts from SummariseOverlaps

We can retrieve the GRanges we counted by the **rowRanges()** accessor function.

The GRanges rows contains our disjoint exons used in counting and is in matching order as rows in our count table.



```r
rowRanges(tregExonCounts)[1:2,]
```

```
## GRanges object with 2 ranges and 3 metadata columns:
##               seqnames               ranges strand |         gene_id
##                  &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;
##   100009600_1     chr9 [21062393, 21062717]      - |       100009600
##   100009600_2     chr9 [21062894, 21062987]      - |       100009600
##                             tx_name exonic_part
##                     &lt;CharacterList&gt;   &lt;integer&gt;
##   100009600_1 uc009veu.1,uc033jjg.1           1
##   100009600_2 uc009veu.1,uc033jjg.1           2
##   -------
##   seqinfo: 66 sequences (1 circular) from mm10 genome
```


---
## Differential transcript usage from RNAseq data.

Previously we have investigated how we may identify changes in the overall abundance on gene.

![](imgs/dge.png)

---
## Differential transcript usage from RNAseq data.

We may also be interested in identifying any evidence of a changes in transcript usage within a gene. 

![](imgs/dtu.png)

---
## Differential transcript usage from RNAseq data.

Due to the uncertainty in the assignment of reads to overlapping transcripts from the same genes, a common approach to investigate differential transcript usage is to define non-overlapping disjoint exons and identify changes in their usage. 


![](imgs/deu2.png)


---
## DEXSeq

Options for differential transcript usage are available in R including DEXseq, and Voom/Limma.

DEXseq can be a little slow when fitting models but it has options for parallization. 

Voom/Limma is significantly faster by comparison but provides a less informative output. 

---
## DEXSeq

To use DEXseq package has a similar set up to that seen for DESeq2.

We first must define our metadata data.frame of sample groups, with sample rownames matching colnames of our **RangedSummarizedExperiment** object. 


```r
metaData &lt;- data.frame(condition=c("Naive","Naive","Act","Act","Act"),
                       row.names = colnames(tregExonCounts))
metaData
```

```
##                        condition
## Sorted_T_reg_act_1.bam     Naive
## Sorted_T_reg_1.bam         Naive
## Sorted_T_reg_act_2.bam       Act
## Sorted_T_reg_2.bam           Act
## Sorted_T_reg_act_3.bam       Act
```


---
## DEXseq

Previously we have been using **DESeqDataSetFromMatrix()** constructor for DESeq2. In DEXSeq we can use the **DEXSeqDataSet** constructor in a similar manner.

As with the **DESeqDataSetFromMatrix()** function, we must provide the counts matrix, our metadata dataframe and our design specifying with metadata columns to test.

In addition to this we must provide our exon IDs and gene IDs to the allow for the association of exons to the genes they constitute.


```r
countMatrix &lt;- assay(tregExonCounts)
countGRanges &lt;- rowRanges(tregExonCounts)
geneIDs &lt;- as.vector(unlist(countGRanges$gene_id))
exonIDs &lt;- rownames(countMatrix)
```



---
## DEXseq


We must specify our design to include a sample and exon effect to distinguish change in exon expression from a change in exon usage.

Previously we simply added a design formula with our group of interest.

```
~ GROUPOFINTERERST
```

Now we specify.

```
~ sample + exon + GROUPOFINTERERST:exon
```

---
## DEXseq

Now we can use the **DEXSeqDataSet** constructor as we have **DESeqDataSetFromMatrix()** but in addition we specify the exon ids to the **featureID** argument and the gene ids to the **groupID** argument.



```r
ddx &lt;- DEXSeqDataSet(countMatrix,metaData,
                     design= ~ sample + exon + condition:exon,
                     featureID = exonIDs,
                     groupID = geneIDs)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 241668 10 
## metadata(1): version
## assays(1): counts
## rownames(241668): 100009600:100009600_1 100009600:100009600_2 ...
##   99982:99982_19 99982:99982_20
## rowData names(4): featureID groupID exonBaseMean exonBaseVar
## colnames: NULL
## colData names(3): sample condition exon
```

---
## DEXseq from  RangedSummarizedExperiment

As we have generated a **RangedSummarizedExperiment** for exon counting, we can use the more straightforward  **DEXSeqDataSetFromSE** function as we used the **DESeqDataSet** function.

We can simply provide our **RangedSummarizedExperiment** and design to the **DEXSeqDataSetFromSE** function.



```r
colData(tregExonCounts)$condition &lt;- c("Act", "Naive", "Act","Naive","Act") 
ddx &lt;- DEXSeqDataSetFromSE(tregExonCounts,
                     design= ~ sample + exon + condition:exon)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 241668 10 
## metadata(1): version
## assays(1): counts
## rownames(241668): 100009600:E001 100009600:E002 ... 99982:E019
##   99982:E020
## rowData names(5): featureID groupID exonBaseMean exonBaseVar
##   transcripts
## colnames: NULL
## colData names(3): sample condition exon
```

---
## DEXseq (filtering low exon counts)

DEXseq is fairly computationally expensive so we can reduce the time taken by removing exons with little expression up front.

Here we can filter any exons which do not have 10 reads in at least two samples (our smallest group size is 2).


```r
ToFilter &lt;- apply(assay(tregExonCounts),1,function(x)sum(x &gt; 10)) &gt; 2
tregExonCounts &lt;- tregExonCounts[ToFilter,]
table(ToFilter)
```

```
## ToFilter
##  FALSE   TRUE 
## 128165 113503
```

```r
ddx &lt;- DEXSeqDataSetFromSE(tregExonCounts,
                     design= ~ sample + exon + condition:exon)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 113503 10 
## metadata(1): version
## assays(1): counts
## rownames(113503): 100017:E001 100017:E002 ... 99982:E019
##   99982:E020
## rowData names(5): featureID groupID exonBaseMean exonBaseVar
##   transcripts
## colnames: NULL
## colData names(3): sample condition exon
```

---
## Running DEXseq in steps.

We can normalize, estimate dispersion and test for differential exon usage with the **estimateSizeFactors()**, **estimateDispersions()** and **testForDEU()** functions. 

Finally we can estimate fold changes for exons using the **estimateExonFoldChanges()** function and specifying metadata column of interest to **fitExpToVar** parameter.


```r
dxd &lt;- estimateSizeFactors(dxd)
dxd &lt;- estimateDispersions(dxd)
dxd &lt;- testForDEU(dxd, reducedModel = ~ sample + exon)
dxd = estimateExonFoldChanges( dxd, fitExpToVar="condition")
```


---
## DEXseq results.

Once have processed our data through the **DEXseq** function we can use the **DEXSeqResults()** function with our **DEXseq** object. We order table to arrange by pvalue.




```r
dxr1 &lt;- DEXSeqResults(dxd)
dxr1 &lt;- dxr1[order(dxr1$pvalue),]
```

---
## DEXseq function.

As with DESeq2, the DEXseq has a single workflow function which will normalise to library sizes, estimate dispersion and perform differential exon usage in one go. The returned object is our **DEXSeqResults** results object.


```r
dxr1 &lt;- DEXSeq(ddx)
dxr1 &lt;- dxr1[order(dxr1$pvalue),]
```

---
## DEXseq results.

The **DEXSeqResults** results object contains significantly more information than the DESeq2 results objects we have seen so far.


```r
dxr1[1,]
```

```
## 
## LRT p-value: full vs reduced
## 
## DataFrame with 1 row and 13 columns
##                groupID   featureID exonBaseMean   dispersion      stat
##            &lt;character&gt; &lt;character&gt;    &lt;numeric&gt;    &lt;numeric&gt; &lt;numeric&gt;
## 18674:E008       18674        E008     1666.058 0.0001130071   606.669
##                  pvalue          padj       Act     Naive
##               &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 18674:E008 5.93257e-134 6.617486e-129  56.89623   70.6842
##            log2fold_Naive_Act               genomicData          countData
##                     &lt;numeric&gt;                 &lt;GRanges&gt;           &lt;matrix&gt;
## 18674:E008          0.7799436 chr10:91122044-91123557:- 1206:2279:1233:...
##            transcripts
##                 &lt;list&gt;
## 18674:E008  uc007gto.1
```

---
## DEXseq results.

The **DEXSeqResults** results object can be converted to a data.frame as with **DESeqResults** objects


```r
as.data.frame(dxr1)[1,]
```

```
##            groupID featureID exonBaseMean   dispersion    stat
## 18674:E008   18674      E008     1666.058 0.0001130071 606.669
##                  pvalue          padj      Act   Naive log2fold_Naive_Act
## 18674:E008 5.93257e-134 6.617486e-129 56.89623 70.6842          0.7799436
##            genomicData.seqnames genomicData.start genomicData.end
## 18674:E008                chr10          91122044        91123557
##            genomicData.width genomicData.strand
## 18674:E008              1514                  -
##            countData.Sorted_T_reg_act_1.bam countData.Sorted_T_reg_1.bam
## 18674:E008                             1206                         2279
##            countData.Sorted_T_reg_act_2.bam countData.Sorted_T_reg_2.bam
## 18674:E008                             1233                         2344
##            countData.Sorted_T_reg_act_3.bam transcripts
## 18674:E008                             1409  uc007gto.1
```

---
## DEXseq results.

A very useful feature of the **DEXSeqResults** results objects is the ability to visualise the differential exons along our gene using the **plotDEXSeq()** function and providing our **DEXSeqResults** results object and gene of interest to plot.


```r
plotDEXSeq(dxr1,"12445")
```

![](RU_RNAseq_p_3_files/figure-html/da4-1.png)&lt;!-- --&gt;

---
## DEXseq results.

We can specify the parameter **displayTranscripts** to TRUE to see the full gene model.

```r
plotDEXSeq(dxr1,"12445",displayTranscripts=TRUE)
```

![](RU_RNAseq_p_3_files/figure-html/da5-1.png)&lt;!-- --&gt;

---
## DEXseq results.

![](imgs/ccnd3_1.png)

---
## DEXseq results.

![](imgs/ccnd3_2.png)

---
## DEXseq results.

We can produce an HTML report for selected genes using the **DEXSeqHTML** function. By default this will create a file called *testForDEU.html* within a folder *DEXSeqReport* in the current working directory.


```r
DEXSeqHTML(dxr1,"12445")
```

[Link to result here](../../Data/DEXSeqReport/testForDEU.html)

---
## DEXseq results- Gene Q values.

It may be desirable to obtain a pvalue summarising the differential exon usage to a gene level.

We can obtain an adjusted p-value per gene with the **perGeneQValue()** with our **DEXSeqResults** results object.

The result is a vector of adjusted p-values named by their corresponding gene.


```r
geneQ &lt;- perGeneQValue(dxr1)
geneQ[1:10]
```

```
##    100017    100019 100033459 100034361 100036523 100036768 100037258 
## 0.9856546 0.1869389 1.0000000 1.0000000 1.0000000 0.9354365 0.9687548 
## 100037262 100037278 100037282 
## 1.0000000 0.5459821 1.0000000
```

---
## DEXseq results- Gene Q values.

We could merge this back to our full results table to provides exon and gene level statistics in one report.


```r
gQFrame &lt;- as.data.frame(geneQ)
dxDF &lt;- as.data.frame(dxr1)
dxDF &lt;-merge(gQFrame,dxDF,by.x=0,by.y=1,all=TRUE)
dxDF &lt;- dxDF[order(dxDF$geneQ,dxDF$pvalue),]
dxDF[1:2,]
```

```
##        Row.names geneQ featureID exonBaseMean   dispersion     stat
## 27532      18674     0      E008     1666.058 0.0001130071 606.6690
## 105162     75547     0      E010     2461.935 0.0001164338 251.8101
##               pvalue          padj      Act    Naive log2fold_Naive_Act
## 27532  5.932570e-134 6.617486e-129 56.89623 70.68420          0.7799436
## 105162  1.046697e-56  4.131555e-52 69.64014 78.61556          0.4643120
##        genomicData.seqnames genomicData.start genomicData.end
## 27532                 chr10          91122044        91123557
## 105162                 chr7          75608485        75610038
##        genomicData.width genomicData.strand
## 27532               1514                  -
## 105162              1554                  +
##        countData.Sorted_T_reg_act_1.bam countData.Sorted_T_reg_1.bam
## 27532                              1206                         2279
## 105162                             1957                         3149
##        countData.Sorted_T_reg_act_2.bam countData.Sorted_T_reg_2.bam
## 27532                              1233                         2344
## 105162                             1879                         3349
##        countData.Sorted_T_reg_act_3.bam transcripts
## 27532                              1409  uc007gto.1
## 105162                             2168  uc009hwx.2
```


---
## RNAseq with multiple groups.

So far we have only reviewed a two group comparison of differential expression and differential transcript usage.

We can compare multiple groups fairly easily in DESeq2.

---
## RNAseq with multiple groups.

For this we will be using the Tissue RNAseq datasets from Encode. 

I have pre-aligned and counted the Tissue RNAseq datasets and provide a **RangedSummarizedExperiment** object

**../Data/gC_TissueFull.RData**

---
## RNAseq with multiple groups.

I load our precounted Tissues' reads gene model and the loaded **RangedSummarizedExperiment** contains counts from three groups.


```r
load("../../Data/gC_TissueFull.RData")
geneCounts_Tissue
```

```
## class: RangedSummarizedExperiment 
## dim: 24100 6 
## metadata(0):
## assays(1): ''
## rownames(24100): 497097 19888 ... 100040911 170942
## rowData names(0):
## colnames(6): Sorted_Heart_1 Sorted_Heart_2 ... Sorted_Liver_1
##   Sorted_Liver_2
## colData names(0):
```

---
## RNAseq with multiple groups.

We can update the **colData** slot of our **RangedSummarizedExperiment** object to allow us to make use of metadata in DESeq2 later.


```r
colData(geneCounts_Tissue)$Tissue &lt;- c("Heart","Heart",
                                       "Kidney","Kidney",
                                       "Liver","Liver")
```

---
## RNAseq with multiple groups.

We can now set up our **DESeq2** object as we have previously and run the **DESeq** workflow function


```r
dds &lt;- DESeqDataSet(geneCounts_Tissue,design = ~Tissue)
dds &lt;- DESeq(dds)
```

---
## RNAseq with multiple groups.

To extract comparisons now we can simply specify the tissues of interest to the results funtion.


```r
heartVsLiver &lt;- results(dds,c("Tissue","Heart","Liver"))
heartVskidney &lt;- results(dds,c("Tissue","Heart","Kidney"))
heartVskidney
```

```
## log2 fold change (MLE): Tissue Heart vs Kidney 
## Wald test p-value: Tissue Heart vs Kidney 
## DataFrame with 24100 rows and 6 columns
##              baseMean log2FoldChange     lfcSE        stat       pvalue
##             &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;    &lt;numeric&gt;
## 497097       1.709418     -0.1114035 1.9683195 -0.05659826 9.548652e-01
## 19888        6.073255      1.6401703 0.9332174  1.75754359 7.882520e-02
## 20671       64.374826      1.6542972 0.4130024  4.00553853 6.187635e-05
## 27395      352.706140      1.0491202 0.2780073  3.77371523 1.608343e-04
## 18777      517.655886     -1.9976114 0.3406895 -5.86343629 4.533849e-09
## ...               ...            ...       ...         ...          ...
## 434935     0.00000000             NA        NA          NA           NA
## 100041256  0.00000000             NA        NA          NA           NA
## 100039574  0.00000000             NA        NA          NA           NA
## 100040911  0.09745882     -0.4459651 4.9965270 -0.08925502    0.9288792
## 170942    10.24190603     -1.3308086 0.9173729 -1.45067346    0.1468708
##                   padj
##              &lt;numeric&gt;
## 497097    9.786147e-01
## 19888     1.666689e-01
## 20671     2.804627e-04
## 27395     6.811435e-04
## 18777     3.523962e-08
## ...                ...
## 434935              NA
## 100041256           NA
## 100039574           NA
## 100040911           NA
## 170942       0.2734518
```

---
## RNAseq with multiple groups.

To identify genes specifically upregulated in Heart versus other tissues. We can then overlap the upregulated genes from the Heart vs Liver and  Heart vs Kidney comparisons.

First we convert **DESeqResults** objects  into data.frame, remove NAs and extract most interesting columns to us.


```r
heartVsLiverDF &lt;- as.data.frame(heartVsLiver)
heartVskidneyDF &lt;- as.data.frame(heartVskidney)
heartVsLiverDF &lt;- heartVsLiverDF[!is.na(heartVsLiverDF$padj),]
heartVskidneyDF &lt;- heartVskidneyDF[!is.na(heartVskidneyDF$padj),]
heartVsLiverDF &lt;- heartVsLiverDF[,c("log2FoldChange","padj")]
heartVskidneyDF &lt;- heartVskidneyDF[,c("log2FoldChange","padj")]
```

---
## RNAseq with multiple groups.

We can then update the column names and merge our data.frame to have a single table of most useful information.


```r
colnames(heartVskidneyDF) &lt;- paste0("HeartVsKidney","_",colnames(heartVskidneyDF))
colnames(heartVsLiverDF) &lt;- paste0("HeartVsLiver","_",colnames(heartVsLiverDF))
fullTable &lt;- merge(heartVsLiverDF,heartVskidneyDF,by=0)
fullTable[1:2,]
```

```
##   Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 1    100017                  -0.5992113         0.2118830
## 2    100019                  -0.6497866         0.1418237
##   HeartVsKidney_log2FoldChange HeartVsKidney_padj
## 1                    0.1578808          0.8033665
## 2                   -0.1634218          0.7817889
```

---
## RNAseq with multiple groups.

Now we can extract our genes upregulated in Heart in both conditions.


```r
upInHeart &lt;- fullTable$HeartVsLiver_log2FoldChange &gt; 0 &amp;
             fullTable$HeartVsKidney_log2FoldChange &gt; 0 &amp;
             fullTable$HeartVsLiver_padj &lt; 0.05 &amp;
             fullTable$HeartVsKidney_padj &lt; 0.05
upInHeartTable &lt;- fullTable[upInHeart,]
upInHeartTable[1:2,]
```

```
##    Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 15 100038347                    4.950045      1.057860e-49
## 37 100038575                    2.454009      2.156496e-02
##    HeartVsKidney_log2FoldChange HeartVsKidney_padj
## 15                     4.342545       2.003331e-50
## 37                     3.161309       1.038031e-03
```

---
## RNAseq with multiple groups.

We can also make a logical data.frame of whether a gene was upregulated in Heart for both Liver and Kidney comparisons.


```r
forVenn &lt;- data.frame(UpvsLiver=fullTable$HeartVsLiver_log2FoldChange &gt; 0 &amp;
                       fullTable$HeartVsLiver_padj &lt; 0.05,
                     UpvsKidney=fullTable$HeartVsKidney_log2FoldChange &gt; 0 &amp;
                       fullTable$HeartVsKidney_padj &lt; 0.05)
forVenn[1:3,]
```

```
##   UpvsLiver UpvsKidney
## 1     FALSE      FALSE
## 2     FALSE      FALSE
## 3     FALSE       TRUE
```

---
## RNAseq with multiple groups.

We can use **limmas'** **vennDiagram()** to produce a venn diagram of overlap in upregulation in heart vs Liver or Kidney. 


```r
library(limma)
vennDiagram(forVenn)
```

![](RU_RNAseq_p_3_files/figure-html/tissue10-1.png)&lt;!-- --&gt;

---
## RNAseq with multiple groups.

When dealing with multiple groups it may be of interest to identify genes are highly variable between groups but with low variability within groups, akin to an Anova test.

In DESeq2, we can identify genes significantly changing across groups by comparing our models with and without our groups of interest using the **results** function.

- Tissue model

```
~ Tissue
```

- No groups model

```
~ 1
```

---
## RNAseq with multiple groups.

To compare out two models we must set the parameter of **reduced** to our alternative model of no groups. We also set the **test** parameter to **LRT** to allow us to compare the models.


```r
dds2 &lt;- DESeq(dds,test="LRT",reduced=~1)
acrossGroups &lt;- results(dds2)
acrossGroups &lt;- acrossGroups[order(acrossGroups$pvalue),]
acrossGroups[1:3,]
```

```
## log2 fold change (MLE): Tissue Liver vs Heart 
## LRT p-value: '~ Tissue' vs '~ 1' 
## DataFrame with 3 rows and 6 columns
##        baseMean log2FoldChange     lfcSE      stat        pvalue
##       &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;
## 16483  125680.4      -2.294864 1.5830283  1642.173  0.000000e+00
## 17888   43583.9     -14.874592 0.6524674  1566.142  0.000000e+00
## 21956   16505.0     -15.058827 1.0578142  1430.916 1.908172e-311
##                padj
##           &lt;numeric&gt;
## 16483  0.000000e+00
## 17888  0.000000e+00
## 21956 1.095545e-307
```


---
## RNAseq with multiple groups.

We can use the **plotCounts()** to get review expression profile of a gene, one at a time. We define the gene of interest to **gene** parameter and **intgroup** to specify metadata column to group counts by.


```r
plotCounts(dds2,gene="17888",intgroup = "Tissue")
```

![](RU_RNAseq_p_3_files/figure-html/tissue12-1.png)&lt;!-- --&gt;

---
## RNAseq with multiple groups.

We can combine our result with our previous table of pair-wise contrasts to provide information on where differences between groups exist.


```r
acrossGroups &lt;- as.data.frame(acrossGroups[,c("pvalue","padj")])
colnames(acrossGroups) &lt;- paste0("AllGroups","_",colnames(acrossGroups))
allGroups &lt;- merge(fullTable,acrossGroups,by.x=1,by.y=0,all=TRUE)
allGroups &lt;- allGroups[order(allGroups$AllGroups_pvalue),]
allGroups[1:2,]
```

```
##      Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 5257     16483                    2.294864      2.469609e-01
## 6057     17888                   14.874592     1.633323e-111
##      HeartVsKidney_log2FoldChange HeartVsKidney_padj AllGroups_pvalue
## 5257                    -16.69262      2.113510e-140                0
## 6057                     14.84458      2.139909e-218                0
##      AllGroups_padj
## 5257              0
## 6057              0
```


---

# Time for an exercise.

[Link_to_exercises](../../Exercises/RNAseq_part3_exercises.html)

[Link_to_answers](../../Answers/RNAseq_part3_answers.html)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
