<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>RU_RNAseq_p4.utf8</title>
    <meta charset="utf-8" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Analysis of RNAseq data in R and Bioconductor (part 4)
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### <a href="http://rockefelleruniversity.github.io/RU_RNAseq/" class="uri">http://rockefelleruniversity.github.io/RU_RNAseq/</a>

---









                                 
## The data

In this session we will be reviewing data from  Jes√∫s Gil's lab at MRC from a splicing factor knockdown (PTBP1 RNAi) RNAseq experiment. Metadata and raw files can be found on GEO [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE101763).

I have aligned all FQ to BAM and counted in genes and exons using Rsubread and summarizeOverlaps().

---
## The data

All Gene and Exon counts can be found as .RData obects in **data** directory

- Counts in disjoint exons can be found at - **data/senescence_ExonCounts.RData**

---
## What we will cover

In our first session we looked at two separate ways we can gain gene expression estimates from raw sequencing data as FASTQ files.

In our second session we imported our expression estimates to identify differences in the expression of genes between two conditions. 

In this session we will look at how we can identify differences in the expression of genes between multiple conditions and how we can identify changes in exon usage to infer changes in transcript usage.

---
## Counts from SummarizeOverlaps

As we saw in session 2, we can use a **BamFileList** to handle multiple BAM files, and **yieldSize** to control memory usage. 


```r
library(Rsamtools)
bamFilesToCount &lt;- c("Sorted_shPTBP1_53_rep1.bam","Sorted_shPTBP1_53_rep2.bam","Sorted_shPTBP1_53_rep3.bam",
                     "Sorted_vec_4OHT_rep1.bam","Sorted_vec_4OHT_rep2.bam","Sorted_vec_4OHT_rep3.bam" )
myBams &lt;- BamFileList(bamFilesToCount,yieldSize = 10000)
```

---
## Counts from SummarizeOverlaps

For differential exon usage we must use a non-overlapping, disjoint set of exons to count on.

We can produce a non-overlapping set of exons by using the **disjointExons()** function. This is equivalent of us having to create GFF file and using the python script wrapped up in DEXseq package (*dexseq_prepare_annotation.py*).


```r
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
nonOverlappingExons &lt;- disjointExons(TxDb.Hsapiens.UCSC.hg19.knownGene)
names(nonOverlappingExons) &lt;- paste(mcols(nonOverlappingExons)$gene_id, mcols(nonOverlappingExons)$exonic_part, 
    sep = "_")
nonOverlappingExons[1:2, ]
```

```
## GRanges object with 2 ranges and 3 metadata columns:
##       seqnames            ranges strand |         gene_id         tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;CharacterList&gt;
##   1_1    chr19 58858172-58858395      - |               1      uc002qsd.4
##   1_2    chr19 58858719-58859006      - |               1      uc002qsd.4
##       exonic_part
##         &lt;integer&gt;
##   1_1           1
##   1_2           2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Disjoint Exons

![](imgs/collapsedE1.png)

---
## Disjoint Exons

![](imgs/collapsedE2.png)


---
## Counts from SummarizeOverlaps

Now we have our disjoint exons to count on we can use the **summarizeOverlaps()**.



```r
library(GenomicAlignments)
senescence_ExonCounts &lt;- summarizeOverlaps(nonOverlappingExons, myBam, ignore.strand = TRUE, 
    inter.feature = FALSE)
senescence_ExonCounts
```


```
## class: RangedSummarizedExperiment 
## dim: 269121 6 
## metadata(0):
## assays(1): counts
## rownames(269121): 1_1 1_2 ... 9997_3 9997_4
## rowData names(3): gene_id tx_name exonic_part
## colnames(6): Sorted_shPTBP1_53_rep1.bam Sorted_shPTBP1_53_rep2.bam ...
##   Sorted_vec_4OHT_rep2.bam Sorted_vec_4OHT_rep3.bam
## colData names(1): condition
```

---
## Counts from SummarizeOverlaps

Our **RangedSummarizedExperiment** contains our counts as a matrix accessible by the **assay()** function.

Each row is named by our Exon ID, geneID + exon number.



```r
assay(senescence_ExonCounts)[1:2, ]
```

```
##     Sorted_shPTBP1_53_rep1.bam Sorted_shPTBP1_53_rep2.bam
## 1_1                        400                        489
## 1_2                        573                        970
##     Sorted_shPTBP1_53_rep3.bam Sorted_vec_4OHT_rep1.bam
## 1_1                        317                      326
## 1_2                        455                      452
##     Sorted_vec_4OHT_rep2.bam Sorted_vec_4OHT_rep3.bam
## 1_1                      650                      508
## 1_2                      801                      885
```
---
## Counts from SummarizeOverlaps

We can retrieve the GRanges we counted by the **rowRanges()** accessor function.

The GRanges rows contains our disjoint exons used in counting and is in matching order as rows in our count table.



```r
rowRanges(senescence_ExonCounts)[1:5, ]
```

```
## GRanges object with 5 ranges and 3 metadata columns:
##       seqnames            ranges strand |         gene_id               tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;       &lt;CharacterList&gt;
##   1_1    chr19 58858172-58858395      - |               1            uc002qsd.4
##   1_2    chr19 58858719-58859006      - |               1            uc002qsd.4
##   1_3    chr19 58859832-58860494      - |               1            uc002qsf.2
##   1_4    chr19 58860934-58861735      - |               1            uc002qsf.2
##   1_5    chr19 58861736-58862017      - |               1 uc002qsd.4,uc002qsf.2
##       exonic_part
##         &lt;integer&gt;
##   1_1           1
##   1_2           2
##   1_3           3
##   1_4           4
##   1_5           5
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
class: inverse, center, middle

# Differential transcript usage

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---
## Differential transcript usage

Previously we have investigated how we may identify changes in the overall abundance on gene using RNAseq data.

![](imgs/dge.png)

---
## Differential transcript usage 

We may also be interested in identifying any evidence of a changes in transcript usage within a gene. 

![](imgs/dtu.png)

---
## Differential transcript usage

Due to the uncertainty in the assignment of reads to overlapping transcripts from the same genes, a common approach to investigate differential transcript usage is to define non-overlapping disjoint exons and identify changes in their usage. 


![](imgs/deu2.png)

---
class: inverse, center, middle

# DEXSeq

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## DEXSeq

Options for differential transcript usage are available in R including DEXseq and Voom/Limma.

DEXseq can be a little slow when fitting models but it has options for parallelization. 

Voom/Limma is significantly faster by comparison, but provides a less informative output. 

---
## DEXSeq inputs

The [**DEXseq** package](https://bioconductor.org/packages/release/bioc/html/DEXSeq.html) has a similar set up to DESeq2.

We first must define our metadata data.frame of sample groups, with sample rownames matching colnames of our **RangedSummarizedExperiment** object. 


```r
metaData &lt;- data.frame(condition = c("shPTBP1_53", "shPTBP1_53", "shPTBP1_53", "senescence", 
    "senescence", "senescence"), row.names = colnames(senescence_ExonCounts))
metaData
```

```
##                             condition
## Sorted_shPTBP1_53_rep1.bam shPTBP1_53
## Sorted_shPTBP1_53_rep2.bam shPTBP1_53
## Sorted_shPTBP1_53_rep3.bam shPTBP1_53
## Sorted_vec_4OHT_rep1.bam   senescence
## Sorted_vec_4OHT_rep2.bam   senescence
## Sorted_vec_4OHT_rep3.bam   senescence
```


---
## DEXseq inputs

Previously we have been using **DESeqDataSetFromMatrix()** constructor for DESeq2. In DEXSeq we can use the **DEXSeqDataSet** constructor in a similar manner.

As with the **DESeqDataSetFromMatrix()** function, we must provide the counts matrix, our metadata dataframe and our design specifying with metadata columns to test.

In addition to this we must provide our exon IDs and gene IDs to the allow for the association of exons to the genes they constitute.


```r
countMatrix &lt;- assay(senescence_ExonCounts)
countGRanges &lt;- rowRanges(senescence_ExonCounts)
geneIDs &lt;- as.vector(unlist(countGRanges$gene_id))
exonIDs &lt;- rownames(countMatrix)
```



---
## DEXseq inputs

We must specify our design to include a sample and exon effect to distinguish change in exon expression from a change in exon usage.

Previously we simply added a design formula with our group of interest.

```
~ GROUPOFINTEREST
```

Now we specify.

```
~ sample + exon + GROUPOFINTEREST:exon
```

---
## Running DEXseq

Now we can use the **DEXSeqDataSet** constructor as we have **DESeqDataSetFromMatrix()** but in addition we specify the exon ids to the **featureID** argument and the gene ids to the **groupID** argument.



```r
library(DEXSeq)
ddx &lt;- DEXSeqDataSet(countMatrix, metaData, design = ~sample + exon + condition:exon, 
    featureID = exonIDs, groupID = geneIDs)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 269121 12 
## metadata(1): version
## assays(1): counts
## rownames(269121): 1:1_1 1:1_2 ... 9997:9997_3 9997:9997_4
## rowData names(4): featureID groupID exonBaseMean exonBaseVar
## colnames: NULL
## colData names(3): sample condition exon
```

---
## Other DEXseq inputs

As we have generated a **RangedSummarizedExperiment** for exon counting, we can use the more straightforward  **DEXSeqDataSetFromSE** function as we used the **DESeqDataSet** function.

We can simply provide our **RangedSummarizedExperiment** and design to the **DEXSeqDataSetFromSE** function.




```r
colData(senescence_ExonCounts)$condition &lt;- factor(c("shPTBP1_53", "shPTBP1_53", 
    "shPTBP1_53", "senescence", "senescence", "senescence"))
ddx &lt;- DEXSeqDataSetFromSE(senescence_ExonCounts, design = ~sample + exon + condition:exon)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 269121 12 
## metadata(1): version
## assays(1): counts
## rownames(269121): 1:E001 1:E002 ... 9997:E003 9997:E004
## rowData names(5): featureID groupID exonBaseMean exonBaseVar
##   transcripts
## colnames: NULL
## colData names(3): sample condition exon
```

---
## DEXseq (filtering low exon counts)

DEXseq is fairly computationally expensive so we can reduce the time taken by removing exons with little expression up front.

Here we can filter any exons which do not have 10 reads in at least two samples (our smallest group size is 3).


```r
ToFilter &lt;- apply(assay(senescence_ExonCounts), 1, function(x) sum(x &gt; 10)) &gt;= 2
senescence_ExonCounts &lt;- senescence_ExonCounts[ToFilter, ]
table(ToFilter)
```

```
## ToFilter
##  FALSE   TRUE 
## 108125 160996
```

```r
ddx &lt;- DEXSeqDataSetFromSE(senescence_ExonCounts, design = ~sample + exon + condition:exon)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 160996 12 
## metadata(1): version
## assays(1): counts
## rownames(160996): 1:E001 1:E002 ... 9997:E002 9997:E003
## rowData names(5): featureID groupID exonBaseMean exonBaseVar
##   transcripts
## colnames: NULL
## colData names(3): sample condition exon
```

---
## Running DEXseq in steps

We can normalize, estimate dispersion and test for differential exon usage with the **estimateSizeFactors()**, **estimateDispersions()** and **testForDEU()** functions. 

Finally we can estimate fold changes for exons using the **estimateExonFoldChanges()** function and specifying metadata column of interest to **fitExpToVar** parameter.


```r
ddx &lt;- estimateSizeFactors(ddx)
ddx &lt;- estimateDispersions(ddx)
ddx &lt;- testForDEU(ddx, reducedModel = ~sample + exon)
ddx = estimateExonFoldChanges(ddx, fitExpToVar = "condition")
```


---
## DEXseq results

Once have processed our data through the **DEXseq** function we can use the **DEXSeqResults()** function with our **DEXseq** object. We order table to arrange by pvalue.




```r
dxr1 &lt;- DEXSeqResults(ddx)
dxr1 &lt;- dxr1[order(dxr1$pvalue), ]
```




---
## DEXseq function

As with DESeq2, the DEXseq has a single workflow function which will normalize to library sizes, estimate dispersion and perform differential exon usage in one go. The returned object is our **DEXSeqResults** results object.


```r
dxr1 &lt;- DEXSeq(ddx)
dxr1 &lt;- dxr1[order(dxr1$pvalue), ]
```

---
## DEXseq results

The **DEXSeqResults** results object contains significantly more information than the DESeq2 results objects we have seen so far.


```r
dxr1[1, ]
```

```
## 
## LRT p-value: full vs reduced
## 
## DataFrame with 1 row and 13 columns
##                groupID   featureID exonBaseMean dispersion      stat    pvalue
##            &lt;character&gt; &lt;character&gt;    &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 57142:E008       57142        E008      21482.1 0.00455953   2036.23         0
##                 padj senescence shPTBP1_53 log2fold_shPTBP1_53_senescence
##            &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;                      &lt;numeric&gt;
## 57142:E008         0    30.1855    63.8085                        4.12122
##                         genomicData             countData           transcripts
##                           &lt;GRanges&gt;              &lt;matrix&gt;                &lt;list&gt;
## 57142:E008 chr2:55252222-55254621:- 40618:35656:54755:... uc002ryd.3,uc002rye.3
```

---
## DEXseq results

The **DEXSeqResults** results object can be converted to a data.frame as with **DESeqResults** objects


```r
as.data.frame(dxr1)[1, ]
```

```
##            groupID featureID exonBaseMean dispersion     stat pvalue padj
## 57142:E008   57142      E008     21482.14 0.00455953 2036.225      0    0
##            senescence shPTBP1_53 log2fold_shPTBP1_53_senescence
## 57142:E008   30.18552   63.80855                       4.121217
##            genomicData.seqnames genomicData.start genomicData.end
## 57142:E008                 chr2          55252222        55254621
##            genomicData.width genomicData.strand
## 57142:E008              2400                  -
##            countData.Sorted_shPTBP1_53_rep1.bam
## 57142:E008                                40618
##            countData.Sorted_shPTBP1_53_rep2.bam
## 57142:E008                                35656
##            countData.Sorted_shPTBP1_53_rep3.bam
## 57142:E008                                54755
##            countData.Sorted_vec_4OHT_rep1.bam
## 57142:E008                               1007
##            countData.Sorted_vec_4OHT_rep2.bam
## 57142:E008                                747
##            countData.Sorted_vec_4OHT_rep3.bam  transcripts
## 57142:E008                                838 uc002ryd....
```

---
## Visualizing DEXseq results

A very useful feature of the **DEXSeqResults** results objects is the ability to visualise the differential exons along our gene using the **plotDEXSeq()** function and providing our **DEXSeqResults** results object and gene of interest to plot.


```r
plotDEXSeq(dxr1, "57142")
```

![](RU_RNAseq_p4_files/figure-html/da4-1.png)&lt;!-- --&gt;

---
## Visualizing DEXseq results

We can specify the parameter **displayTranscripts** to TRUE to see the full gene model.

```r
plotDEXSeq(dxr1, "57142", displayTranscripts = TRUE)
```

![](RU_RNAseq_p4_files/figure-html/da5c-1.png)&lt;!-- --&gt;

---
## Visualizing DEXseq results

![](imgs/RTN4.png)



---
## Visualizing DEXseq results

The differential exon usage for Entrez gene 7168 is more complex.


```r
plotDEXSeq(dxr1, "7168", displayTranscripts = TRUE)
```

![](RU_RNAseq_p4_files/figure-html/da5-1.png)&lt;!-- --&gt;

---
## Visualizing DEXseq results


![](imgs/TPM1.png)

---
## Visualizing DEXseq results

We can produce an HTML report for selected genes using the **DEXSeqHTML** function. By default this will create a file called *testForDEU.html* within a folder *DEXSeqReport* in the current working directory.


```r
DEXSeqHTML(dxr1, "57142")
```

[Link to result here]("../../data/DEXSeqReport/testForDEU.html")

---
## Gene Q values

It may be desirable to obtain a pvalue summarizing the differential exon usage to a gene level.

We can obtain an adjusted p-value per gene with the **perGeneQValue()** with our **DEXSeqResults** results object.

The result is a vector of adjusted p-values named by their corresponding gene.


```r
geneQ &lt;- perGeneQValue(dxr1)
geneQ[1:10]
```

```
##         1       100      1000     10000     10001 100033414 100037417 100048912 
##         1         1         1         1         1         1         1         1 
## 100049076     10005 
##         1         1
```

---
## Gene Q values

We could merge this back to our full results table to provides exon and gene level statistics in one report.


```r
gQFrame &lt;- as.data.frame(geneQ, row.names = names(geneQ))
dxDF &lt;- as.data.frame(dxr1)
dxDF &lt;- merge(gQFrame, dxDF, by.x = 0, by.y = 1, all = TRUE)
dxDF &lt;- dxDF[order(dxDF$geneQ, dxDF$pvalue), ]
dxDF[1:2, ]
```

```
##        Row.names geneQ featureID exonBaseMean  dispersion     stat
## 97377      57142     0      E008    21482.144 0.004559530 2036.225
## 117732      7168     0      E009     2805.899 0.004776165 1092.222
##               pvalue          padj senescence shPTBP1_53
## 97377   0.000000e+00  0.000000e+00   30.18552   63.80855
## 117732 1.619401e-239 1.246016e-234   49.34291   18.36318
##        log2fold_shPTBP1_53_senescence genomicData.seqnames genomicData.start
## 97377                        4.121217                 chr2          55252222
## 117732                      -4.264727                chr15          63353397
##        genomicData.end genomicData.width genomicData.strand
## 97377         55254621              2400                  -
## 117732        63353472                76                  +
##        countData.Sorted_shPTBP1_53_rep1.bam
## 97377                                 40618
## 117732                                  619
##        countData.Sorted_shPTBP1_53_rep2.bam
## 97377                                 35656
## 117732                                  668
##        countData.Sorted_shPTBP1_53_rep3.bam countData.Sorted_vec_4OHT_rep1.bam
## 97377                                 54755                               1007
## 117732                                  705                               4442
##        countData.Sorted_vec_4OHT_rep2.bam countData.Sorted_vec_4OHT_rep3.bam
## 97377                                 747                                838
## 117732                               4979                               5206
##         transcripts
## 97377  uc002ryd....
## 117732 uc002alh....
```


---
class: inverse, center, middle

# RNAseq with multiple groups

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## RNAseq with multiple groups

So far we have only reviewed a two group comparison of differential expression and differential transcript usage.

We can compare multiple groups fairly easily in DESeq2.

---
## The data

For this we will be using Tissue specific RNAseq datasets from ENCODE. 

I have pre-aligned and counted the Tissue RNAseq datasets and provide a **RangedSummarizedExperiment** object

**data/gC_TissueFull.RData**

---
## The data

When we load in the data object we have a **RangedSummarizedExperiment** which contains counts from three groups.


```r
load("data/gC_TissueFull.RData")
geneCounts_Tissue
```

```
## class: RangedSummarizedExperiment 
## dim: 24100 6 
## metadata(0):
## assays(1): ''
## rownames(24100): 497097 19888 ... 100040911 170942
## rowData names(0):
## colnames(6): Sorted_Heart_1 Sorted_Heart_2 ... Sorted_Liver_1
##   Sorted_Liver_2
## colData names(0):
```

---
## DEseq2 input

We can update the **colData** slot of our **RangedSummarizedExperiment** object to allow us to make use of metadata in DESeq2 later.


```r
colData(geneCounts_Tissue)$Tissue &lt;- c("Heart", "Heart", "Kidney", "Kidney", "Liver", 
    "Liver")
```

---
## DEseq2 input

We can now set up our **DESeq2** object as we have previously and run the **DESeq** workflow function


```r
dds &lt;- DESeqDataSet(geneCounts_Tissue, design = ~Tissue)
dds &lt;- DESeq(dds)
```

---
## Running DEseq2

To extract comparisons now we can simply specify the tissues of interest to the results funtion.


```r
heartVsLiver &lt;- results(dds, c("Tissue", "Heart", "Liver"))
heartVskidney &lt;- results(dds, c("Tissue", "Heart", "Kidney"))
heartVskidney
```

```
## log2 fold change (MLE): Tissue Heart vs Kidney 
## Wald test p-value: Tissue Heart vs Kidney 
## DataFrame with 24100 rows and 6 columns
##             baseMean log2FoldChange     lfcSE       stat      pvalue
##            &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
## 497097       1.70942      -0.111433  1.968710 -0.0566022 9.54862e-01
## 19888        6.07325       1.640170  0.933425  1.7571520 7.88919e-02
## 20671       64.37483       1.654297  0.413053  4.0050438 6.20060e-05
## 27395      352.70614       1.049120  0.278011  3.7736704 1.60863e-04
## 18777      517.65589      -1.997611  0.340689 -5.8634514 4.53344e-09
## ...              ...            ...       ...        ...         ...
## 434935     0.0000000             NA        NA         NA          NA
## 100041256  0.0000000             NA        NA         NA          NA
## 100039574  0.0000000             NA        NA         NA          NA
## 100040911  0.0974588      -0.445965  4.996527  -0.089255    0.928879
## 170942    10.2419060      -1.330817  0.917523  -1.450445    0.146934
##                  padj
##             &lt;numeric&gt;
## 497097    9.79719e-01
## 19888     1.66810e-01
## 20671     2.81050e-04
## 27395     6.81433e-04
## 18777     3.52364e-08
## ...               ...
## 434935             NA
## 100041256          NA
## 100039574          NA
## 100040911          NA
## 170942        0.27357
```

---
## Comparing multiple results

To identify genes specifically upregulated in Heart versus other tissues. We can then overlap the upregulated genes from the Heart vs Liver and  Heart vs Kidney comparisons.

First we convert **DESeqResults** objects into a data.frame, remove NAs and extract most interesting columns to us.


```r
heartVsLiverDF &lt;- as.data.frame(heartVsLiver)
heartVskidneyDF &lt;- as.data.frame(heartVskidney)

heartVsLiverDF &lt;- heartVsLiverDF[!is.na(heartVsLiverDF$padj), ]
heartVskidneyDF &lt;- heartVskidneyDF[!is.na(heartVskidneyDF$padj), ]

heartVsLiverDF &lt;- heartVsLiverDF[, c("log2FoldChange", "padj")]
heartVskidneyDF &lt;- heartVskidneyDF[, c("log2FoldChange", "padj")]
```

---
## Comparing multiple results

We can then update the column names and merge our data.frame to have a single table of most useful information.


```r
colnames(heartVskidneyDF) &lt;- paste0("HeartVsKidney", "_", colnames(heartVskidneyDF))
colnames(heartVsLiverDF) &lt;- paste0("HeartVsLiver", "_", colnames(heartVsLiverDF))
fullTable &lt;- merge(heartVsLiverDF, heartVskidneyDF, by = 0)
fullTable[1:2, ]
```

```
##   Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 1    100017                  -0.5992115         0.2119439
## 2    100019                  -0.6497866         0.1418297
##   HeartVsKidney_log2FoldChange HeartVsKidney_padj
## 1                    0.1578806          0.8036222
## 2                   -0.1634218          0.7819743
```

---
## Comparing multiple results

Now we can extract our genes upregulated in Heart in both conditions.


```r
upInHeart &lt;- fullTable$HeartVsLiver_log2FoldChange &gt; 0 &amp; fullTable$HeartVsKidney_log2FoldChange &gt; 
    0 &amp; fullTable$HeartVsLiver_padj &lt; 0.05 &amp; fullTable$HeartVsKidney_padj &lt; 0.05
upInHeartTable &lt;- fullTable[upInHeart, ]
upInHeartTable[1:2, ]
```

```
##    Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 15 100038347                    4.950045      1.057785e-49
## 37 100038575                    2.454005      2.159250e-02
##    HeartVsKidney_log2FoldChange HeartVsKidney_padj
## 15                     4.342545       2.003141e-50
## 37                     3.161305       1.040576e-03
```

---
## Do genes overlap?

We can also make a logical data.frame of whether a gene was upregulated in Heart for both Liver and Kidney comparisons.


```r
forVenn &lt;- data.frame(UpvsLiver = fullTable$HeartVsLiver_log2FoldChange &gt; 0 &amp; fullTable$HeartVsLiver_padj &lt; 
    0.05, UpvsKidney = fullTable$HeartVsKidney_log2FoldChange &gt; 0 &amp; fullTable$HeartVsKidney_padj &lt; 
    0.05)
forVenn[1:3, ]
```

```
##   UpvsLiver UpvsKidney
## 1     FALSE      FALSE
## 2     FALSE      FALSE
## 3     FALSE       TRUE
```

---
## Do genes overlap?

We can use **limmas'** **vennDiagram()** to produce a venn diagram of overlap in upregulation in heart vs Liver or Kidney. 


```r
library(limma)
vennDiagram(forVenn)
```

![](RU_RNAseq_p4_files/figure-html/tissue10-1.png)&lt;!-- --&gt;

---
## RNAseq with multiple groups

When dealing with multiple groups it may be of interest to identify genes are highly variable between groups but with low variability within groups, akin to an Anova test.

In DESeq2, we can identify genes significantly changing across groups by comparing our models with and without our groups of interest using the **results** function.

- Tissue model

```
~ Tissue
```

- No groups model

```
~ 1
```

---
## Likelihood-ratio test

To compare out two models we must set the parameter of **reduced** to our alternative model of no groups. We also set the **test** parameter to **LRT** to allow us to compare the models.


```r
dds2 &lt;- DESeq(dds, test = "LRT", reduced = ~1)
acrossGroups &lt;- results(dds2)
acrossGroups &lt;- acrossGroups[order(acrossGroups$pvalue), ]
acrossGroups[1:3, ]
```

```
## log2 fold change (MLE): Tissue Liver vs Heart 
## LRT p-value: '~ Tissue' vs '~ 1' 
## DataFrame with 3 rows and 6 columns
##        baseMean log2FoldChange     lfcSE      stat       pvalue         padj
##       &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;
## 16483  125680.4       -2.29486  1.583024   1642.43  0.00000e+00  0.00000e+00
## 17888   43583.9      -14.87459  0.652461   1566.29  0.00000e+00  0.00000e+00
## 21956   16505.0      -15.05883  1.057810   1431.05 1.78171e-311 1.02294e-307
```


---
## Likelihood-ratio test

We can use the **plotCounts()** to get review expression profile of a gene, one at a time. We define the gene of interest to **gene** parameter and **intgroup** to specify metadata column to group counts by.


```r
plotCounts(dds2, gene = "17888", intgroup = "Tissue")
```

![](RU_RNAseq_p4_files/figure-html/tissue12-1.png)&lt;!-- --&gt;

---
## RNAseq with multiple groups

We can combine our result with our previous table of pair-wise contrasts to provide information on where differences between groups exist.


```r
acrossGroups &lt;- as.data.frame(acrossGroups[, c("pvalue", "padj")])
colnames(acrossGroups) &lt;- paste0("AllGroups", "_", colnames(acrossGroups))
allGroups &lt;- merge(fullTable, acrossGroups, by.x = 1, by.y = 0, all = TRUE)
allGroups &lt;- allGroups[order(allGroups$AllGroups_pvalue), ]
allGroups[1:2, ]
```

```
##      Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 5257     16483                    2.294861      2.469712e-01
## 6057     17888                   14.874592     1.625126e-111
##      HeartVsKidney_log2FoldChange HeartVsKidney_padj AllGroups_pvalue
## 5257                    -16.69262      2.090655e-140                0
## 6057                     14.84458      2.099275e-218                0
##      AllGroups_padj
## 5257              0
## 6057              0
```


---

# Time for an exercise

[Link_to_exercises](../../exercises/exercises/RNAseq_part4_exercise.html)

[Link_to_answers](../../exercises/answers/RNAseq_part4_answers.html)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
