---
title: "RNA-seq in Bioconductor exercises"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_RNAseq/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---


```{r setup, include=FALSE,message=FALSE,warning=FALSE}
require(DESeq2)
```

In todays session we will work with some of the RNA-seq data of adult mouse tissues from Bing Ren's lab, Liver and Heart. 
- More information on liver data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CHB/)
- More information on heart data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)
- More information on Kidney data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)

- Precounted RNAseq reads in genes for these tissues can be found as an R data object in **/Data/gC_TissueFull.RData**.

## DGE RNA-seq analysis.

1. Load the **/Data/gC_TissueFull.RData** and create a DESeq2 object. Create a venn diagram of genes upregulated in Kidney versus Liver and Heart.

```{r, include=FALSE,warning=FALSE,echo=toMessage}
library(DESeq2)
library(tximport)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(goseq)
library(KEGG.db)
library(ggplot2)
```

```{r a,cache=TRUE,eval=TRUE,warning=FALSE,echo=toMessage}
library(DESeq2)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)

load("data/gC_TissueFull.RData")

colData(geneCounts_Tissue)$Tissue <- factor(c("Heart","Heart",
                                              "Kidney","Kidney",
                                              "Liver","Liver"))

ddsT <- DESeqDataSet(geneCounts_Tissue,design=~Tissue)
ddsT <- DESeq(ddsT)

KidneyVsLiver <- results(ddsT,c("Tissue","Kidney","Liver"))
KidneyVsHeart <- results(ddsT,c("Tissue","Kidney","Heart"))
KidneyVsLiverDF <- as.data.frame(KidneyVsLiver)
KidneyVsHeartDF <- as.data.frame(KidneyVsHeart)
KidneyVsLiverDF <- KidneyVsLiverDF[!is.na(KidneyVsLiverDF$padj),]
KidneyVsHeartDF <- KidneyVsHeartDF[!is.na(KidneyVsHeartDF$padj),]
KidneyVsLiverDF <- KidneyVsLiverDF[,c("log2FoldChange","padj")]
KidneyVsHeartDF <- KidneyVsHeartDF[,c("log2FoldChange","padj")]
colnames(KidneyVsHeartDF) <- paste0("KidneyVsHeart","_",colnames(KidneyVsHeartDF))
colnames(KidneyVsLiverDF) <- paste0("KidneyVsLiver","_",colnames(KidneyVsLiverDF))
fullTable <- merge(KidneyVsLiverDF,KidneyVsHeartDF,by=0)

forVenn <- data.frame(UpvsLiver=fullTable$KidneyVsLiver_log2FoldChange > 0 &
                       fullTable$KidneyVsLiver_padj < 0.05,
                     UpvsHeart=fullTable$KidneyVsHeart_log2FoldChange > 0 &
                       fullTable$KidneyVsHeart_padj < 0.05)
library(limma)
vennDiagram(forVenn)
```

2. Perform a enrichment analysis for genes upregulated in Kidney vs both liver and Heart using KEGG pathways and plot the -log10 pvalue of the top 10 terms as a scatter plot.


```{r b,cache=TRUE,eval=TRUE,dependson="a",warning=FALSE,echo=toMessage}
rownames(forVenn) <- fullTable$Row.names
UpvsBoth <-  rowSums(forVenn) == 2
library(goseq)
pwf = nullp(UpvsBoth, "mm10", "knownGene", plot.fit = TRUE)
KEGG_UpInKindey <- goseq(pwf,"mm10","knownGene",
                       test.cats=c("KEGG"))

library(KEGG.db)
xx <- as.list(KEGGPATHID2NAME)
idtoName <- cbind(names(xx),unlist(xx))
KEGG_UpInKindey <- merge(idtoName,KEGG_UpInKindey,by=1,all=FALSE)
KEGG_UpInKindey <- KEGG_UpInKindey[order(KEGG_UpInKindey$over_represented_pvalue),]
ggplot(KEGG_UpInKindey[1:10,],aes(x=-log10(over_represented_pvalue),
                                         size=numDEInCat,y=V2))+geom_point()+theme_bw()
```

3. Create a boxplot plot of log2FC of genes (as from Kidney vs Liver and Kidney vs Heart comparison) in collecting duct acid secretion term vs all genes not in duct acid secretion term.

```{r c,cache=TRUE,eval=TRUE,dependson="b",warning=FALSE,echo=toMessage}

library(org.Mm.eg.db)
cdaSecretion <- AnnotationDbi::select(org.Mm.eg.db,keytype = "PATH",
                                      keys = "04966",columns = "ENTREZID")

kLF <- data.frame(Log2FC=fullTable$KidneyVsLiver_log2FoldChange,
           Comparison="KidneyVsLiver",
           EntrezID=fullTable$Row.names)

kHF <- data.frame(Log2FC=fullTable$KidneyVsHeart_log2FoldChange,
           Comparison="KidneyVsHeart",
           EntrezID=fullTable$Row.names)
toPlot <- rbind(kLF,kHF)
toPlot$GeneSet <- ifelse(toPlot$EntrezID %in% cdaSecretion$ENTREZID,"cdaSecretion","Other")

ggplot(toPlot,aes(x=Comparison,y=Log2FC,fill=GeneSet))+geom_boxplot()+theme_minimal()
```

4. Identify genes changing significantly in any condition (< 0.05) and produce a file of results containing gene names.

```{r k,cache=TRUE,eval=TRUE,dependson="a",warning=FALSE,echo=toMessage}
library(org.Mm.eg.db)
ddsAll <- DESeq(ddsT,test="LRT",reduced=~1)
myRes <- results(ddsAll)
myResDF <- as.data.frame(myRes)
library(org.Mm.eg.db)
eToSym <- AnnotationDbi::select(org.Mm.eg.db,
                 keys = rownames(myResDF),
                 keytype = "ENTREZID",
                 columns="SYMBOL")
annotatedRes <- merge(eToSym,myResDF,
                      by.x=1,
                      by.y=0,all=FALSE)
annotatedRes <- annotatedRes[order(annotatedRes$pvalue),]
annotatedRes <- annotatedRes[!is.na(annotatedRes$padj),]
write.table(annotatedRes,file="Diff_Reduced.csv",sep=",",row.names=FALSE,quote=FALSE)
```



5. DTU-  Load the DEXseqResults object (found in **../Data/dxr_HeartVsLiver.RData**) and make a plot of the differnetial splicing for the Atp2a2 gene.

```{r e,cache=TRUE,eval=TRUE,warning=FALSE,echo=toMessage}
library(DEXSeq)
load("data/dxr_HeartVsLiver.RData")

dxr1DF <- as.data.frame(dxr1)
dxr1DF <- dxr1DF[order(dxr1DF$pvalue),]
library(org.Mm.eg.db)
eToSym <- AnnotationDbi::select(org.Mm.eg.db,
                 keys = unique(dxr1DF[,1]),
                 keytype = "ENTREZID",
                 columns="SYMBOL")
annotatedRes <- merge(eToSym,dxr1DF,
                      by.x=1,
                      by.y=1,all=FALSE)
annotatedRes <- annotatedRes[order(annotatedRes$pvalue),]
annotatedRes[1:3,]
plotDEXSeq(dxr1,11938,fitExpToVar = "tissue",displayTranscripts = TRUE)
```


6. DTU-  Load the RangedSummarizedExperiment object (found in **../Data/RSE_HeartAndLiver.RData** - obect is tissueExonCounts) and perform a differential exon analysis. Make a plot of the gene containing the most significantly differentially used exon.

```{r f,cache=TRUE,eval=TRUE,warning=FALSE,echo=toMessage}
load(file="data/RSE_HeartAndLiver.RData")

ddxTissue <- DEXSeqDataSetFromSE(tissueExonCounts,
                      design= ~ sample + exon + tissue:exon)

ddxTissue <- DEXSeq(ddxTissue,fitExpToVar = "tissue")
ddxTissue <- ddxTissue[order(ddxTissue$pvalue),]
plotDEXSeq(ddxTissue,16561,fitExpToVar = "tissue",displayTranscripts = TRUE)
```