---
title: "RNAseq in Bioconductor exercises (part 5)"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_RNAseq/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r setup, include=FALSE,message=FALSE,warning=FALSE}
require(DESeq2)
```

In todays session we will work with some of the RNAseq data of adult mouse tissues from Bing Ren's lab, Liver and Heart. 

- More information on liver data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CHB/)
- More information on heart data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)
- More information on Kidney data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)

- Precounted RNAseq reads in genes for these tissues can be found as an R data object in **data/gC_TissueFull.RData**.


## Exercises

### 1. Differential Transcript Usage
We have run the DEXseq workflow to compare Heart and Liver. Load the DEXseqResults object called dxr1 (found in **data/dxr_HeartVsLiver.RData**) and make a plot of the differential splicing for the Atp2a2 gene.

```{r e,cache=TRUE,eval=TRUE,warning=FALSE,echo=toMessage}
library(DEXSeq)
load("data/dxr_HeartVsLiver.RData")

dxr1DF <- as.data.frame(dxr1)
dxr1DF <- dxr1DF[order(dxr1DF$pvalue),]
library(org.Mm.eg.db)
eToSym <- AnnotationDbi::select(org.Mm.eg.db,
                 keys = unique(dxr1DF[,1]),
                 keytype = "ENTREZID",
                 columns="SYMBOL")
annotatedRes <- merge(eToSym,dxr1DF,
                      by.x=1,
                      by.y=1,all=FALSE)
annotatedRes <- annotatedRes[order(annotatedRes$pvalue),]
annotatedRes[1:3,]
plotDEXSeq(dxr1,11938,fitExpToVar = "tissue",displayTranscripts = TRUE)
```


### 2. Differential Transcript Usage
Load the RangedSummarizedExperiment object called tissueExonCounts (found in **data/RSE_HeartAndLiver.RData**). Perform a differential exon analysis. Make a plot of the gene containing the most significantly differentially used exon.

```{r f,cache=TRUE,eval=TRUE,warning=FALSE,echo=toMessage}
load(file="data/RSE_HeartAndLiver.RData")

ddxTissue <- DEXSeqDataSetFromSE(tissueExonCounts,
                      design= ~ sample + exon + tissue:exon)

ddxTissue <- DEXSeq(ddxTissue,fitExpToVar = "tissue")
ddxTissue <- ddxTissue[order(ddxTissue$pvalue),]
plotDEXSeq(ddxTissue,16561,fitExpToVar = "tissue",displayTranscripts = TRUE)
```