---
title: "RNAseq in Bioconductor exercises (part 4)"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_RNAseq/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---


```{r setup, include=FALSE}
library(DESeq2)
```

In todays session we will work with some of the RNAseq data of adult mouse tissues from Bing Ren's lab, Liver and Heart.  

- More information on liver data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CHB/)
- More information on heart data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)

- Differential expression results for Heart minus Liver can be found in the data directory:  

* DGE  - **data/Heart_minus_liver.csv**
* Counts - **data/gC_TissueFull.RData**

## Exercises

### 1. *goseq* cellular component analysis

Identify GO term cellular component groups enriched in genes significantly upregulated in Liver with goseq. What are the top 5 terms?

*HINT: Use "GO:CC" as the test category for goseq*


```{r ea,cache=TRUE,eval=F,dependson="d",warning=FALSE,echo=toMessage}
library(clusterProfiler)

newResDF <- read.delim("data/Heart_minus_liver.csv",sep=",")
newResDF <- newResDF[!is.na(newResDF$padj),]

UpInLiver <- newResDF$padj < 0.05 & 
             newResDF$log2FoldChange < 0
UpInLiver <- as.integer(UpInLiver)
names(UpInLiver) <- newResDF$ID

library(goseq)
pwf = nullp(UpInLiver, "mm10", "knownGene", plot.fit = TRUE)
CC_UpInLiver <- goseq(pwf,"mm10","knownGene",
                       test.cats=c("GO:CC"))

CC_UpInLiver[1:5,]
```


<!-- ```{r} -->
<!-- # library(KEGG.db) -->
<!-- # xx <- as.list(KEGGPATHID2NAME) -->
<!-- # idtoName <- cbind(names(xx),unlist(xx)) -->
<!-- #  -->
<!-- # Kegg_UpInLiver <- merge(idtoName,Kegg_UpInLiver,by=1,all=TRUE) -->
<!-- # orderByP <- order(Kegg_UpInLiver$over_represented_pvalue) -->
<!-- # Kegg_UpInLiver <- Kegg_UpInLiver[orderByP,] -->
<!-- ``` -->

### 2. ClusterProfiler for KEGG enrichment
Plot the -log10 pvalue for top 5 terms as point plot.

**HINT: The "gseKEGG" *
```{r f,cache=TRUE,eval=F,dependson="e",warning=FALSE,echo=toMessage}
library(clusterProfiler)
geneList <- newResDF$stat
names(geneList) <-  newResDF$ID
geneList <- geneList[order(geneList, decreasing = T)]


KEGG_HeartLiver <- gseKEGG(
  geneList,
  organism = "mmu",
  keyType = "kegg", eps = 10e-20)

head(KEGG_HeartLiver)
```

### 3. Dotplot
Show dotplot of the clusterProfiler results. Show the top 20 terms. 

```{r g,cache=TRUE,eval=FALSE,warning=FALSE,echo=toMessage}
dotplot(KEGG_HeartLiver, showCategory = 20)

```

### 4. Enrichment Map
Show emap of the clusterProfiler results. Show the top 20 terms. 

```{r,cache=TRUE,eval=FALSE,warning=FALSE,echo=toMessage}
library(enrichplot)
KEGG_HeartLiver_pt <- pairwise_termsim(KEGG_HeartLiver)

emapplot(KEGG_HeartLiver_pt,cex_label_category=0.5) 

```


### 5. Heatmap of Leading Edge Genes
Draw a heatmap of the genes driving enrichment in the top 3 KEGG terms. Use rlog counts, scale across row using a Z-score and include the kidney data as a reference point. 

*HINT - The gene IDs are in the core_enrichment column of the clusterProfiler result*

```{r,echo=toMessage}
library(pheatmap)
library(DESeq2)

my_gene_IDs <- unlist(strsplit(KEGG_HeartLiver$core_enrichment[1],"/"))

load("data/gC_TissueFull.RData")

dds <- DESeqDataSet(geneCounts, design = ~Tissue)
dds <- DESeq(dds)

rlogTissue <- rlog(dds)


mat_for_heatmap <- assay(rlogTissue)[rownames(rlogTissue) %in% my_gene_IDs,]

annoDF <- as.data.frame(colData(dds))
annoDF <- annoDF[,1,drop=F]


pheatmap(mat_for_heatmap,
         scale="row", show_rownames = F,
         annotation_col = annoDF)


```

