---
title: "RNAseq in Bioconductor exercises (part 1)"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_RNAseq/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
```


<p>&nbsp;</p>
These are the first exercises are about alignment and counting in [RNAseq](https://rockefelleruniversity.github.io/RU_RNAseq/).

In todays session we will work with some of the RNAseq data of T-regulatory cells from Christina Leslie's lab.

Sequencing data as a FASTQ file [can be found here.](https://www.encodeproject.org/files/ENCFF070QMF/@@download/ENCFF070QMF.fastq.gz)

Aligned data as a BAM file [can be found here.](https://www.encodeproject.org/files/ENCFF906UTB/@@download/ENCFF906UTB.bam)


## Exercises

### 1. Basic QC metrics
Download the above FASTQ file for T-regulatory cells (replicate 2) - **ENCFF070QMF.fastq.gz**. Sample 10000 reads from file and produce our 3 standard plots: 
  -  DNA base occurrence over cycles
  -  histogram of quality scores 
  -  Boxplot of quality scores

```{r, include=FALSE,echo=toMessage}
library(ShortRead)
library(Rsamtools)
library(BSgenome.Mmusculus.UCSC.mm10)

# if(!file.exists("~/ENCFF070QMF.fastq.gz")){
#   download.file("https://www.encodeproject.org/files/ENCFF070QMF/@@download/ENCFF070QMF.fastq.gz","~/ENCFF070QMF.fastq.gz")
# }

```

```{r instalcscldcsceBSded,cache=TRUE,eval=F,echo=toMessage}
library(ShortRead)
library(ggplot2)
library(Rsubread)

fqSample <- FastqSampler("ENCFF070QMF.fastq.gz",n=10^5)
fastq <- yield(fqSample)
```

```{r, eval=T, echo=FALSE}
library(ShortRead)
library(ggplot2)
library(Rsubread)
load("data/ENCFF070QMF_sampled.fastq.gz")

```

```{r cache=TRUE,eval=TRUE, echo=toMessage}
fastq

readSequences <- sread(fastq)
readSequences_AlpbyCycle <- alphabetByCycle(readSequences)

AFreq <- readSequences_AlpbyCycle["A",]
CFreq <- readSequences_AlpbyCycle["C",]
GFreq <- readSequences_AlpbyCycle["G",]
TFreq <- readSequences_AlpbyCycle["T",]
toPlot <- data.frame(Count=c(AFreq,CFreq,GFreq,TFreq),
                     Cycle=rep(1:50,4),
                     Base=rep(c("A","C","G","T"),each=50))
ggplot(toPlot,aes(y=Count,x=Cycle,colour=Base))+geom_line()+theme_bw()




readQuality <- quality(fastq)
readQualityScores <- alphabetScore(readQuality)

toPlot <- data.frame(ReadQ=readQualityScores)
ggplot(toPlot,aes(x=ReadQ))+geom_histogram()+theme_minimal()


qualAsMatrix <- as(readQuality,"matrix")
boxplot(qualAsMatrix[1:10000,])
```

### 2. Filter out low quality reads
Remove reads with summed qualities less than 1000 and write to a new FASTQ file.

```{r instaaxscslldebbBSccdxxed,cache=TRUE,eval=TRUE,dependson="instalcscldcsceBSded",echo=toMessage}
fastqFilt <- fastq[readQualityScores >= 1000]

writeFastq(fastqFilt,"data/filtered_ENCFF070QMF.fastq.gz")

```

### 3. Alignment
Align our filtered reads to the chromosome 10 of mm10 genome. Sort and index the resulting BAM file.

```{r instaaxscslldebbzBSccdxxed2,cache=FALSE,eval=F,echo=toMessage}

library(BSgenome.Mmusculus.UCSC.mm10)
library(Rsubread)

mainChromosomes <- paste0("chr10")
mainChrSeq <- lapply(mainChromosomes,
                     function(x)BSgenome.Mmusculus.UCSC.mm10[[x]])
names(mainChrSeq) <- mainChromosomes
mainChrSeqSet <- DNAStringSet(mainChrSeq)
writeXStringSet(mainChrSeqSet,
                "mm10Chr10.fa")



library(TxDb.Mmusculus.UCSC.mm10.knownGene)
myExons <- exons(TxDb.Mmusculus.UCSC.mm10.knownGene,columns=c("tx_id","gene_id"))
myExons <- myExons[lengths(myExons$gene_id) == 1]
dfExons <- as.data.frame(myExons)
SAF <- data.frame(GeneID=dfExons$gene_id,
                  Chr=dfExons$seqnames,
                  Start=dfExons$start,
                  End=dfExons$end,
                  Strand=dfExons$strand)

## Rsubread

buildindex("mm10Chr10",
           "mm10Chr10.fa")
subjunc("mm10Chr10",
      "filtered_ENCFF070QMF.fastq.gz",
      output_file="Treg_2.bam",annot.ext = SAF,isGTF = FALSE,useAnnotation = TRUE)


sortBam("Treg_2.bam","Sorted_Treg_2")
indexBam("Sorted_Treg_2.bam")

```


### 4. Alignment
Count the reads in our newly aligned and indexed BAM file mapping within genes. Plot a density plot of log 10 of reads counts across genes on chromosome 10.  
**NOTE: Add 1 read to all counts to avoid log of zero**


```{r instacsxsjiijcsllBSded,echo=toMessage, eval=F}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
geneExons <- exonsBy(TxDb.Mmusculus.UCSC.mm10.knownGene,by="gene")
library(GenomicAlignments)
myBam <- BamFile("Sorted_Treg_2.bam",yieldSize = 10000)
treg2GeneCounts <- summarizeOverlaps(geneExons,myBam,
                                    ignore.strand = TRUE)

myGeneGR <- rowRanges(treg2GeneCounts)
treg2GeneCountsChr10 <- treg2GeneCounts[all(seqnames(myGeneGR) ==
    "chr10"),]
treg2GeneCountsChr10Matrix <- assay(treg2GeneCountsChr10)+1
myCounts <- data.frame(Counts =treg2GeneCountsChr10Matrix[,1])
ggplot(myCounts,aes(x=Counts))+geom_density(fill="Red")+scale_x_log10()+theme_minimal()
```

### 5. (Optional) Salmon Quantification
Download and install Salmon. Using Salmon, quantify transcript levels using reads from our filtered FASTQ.
```{r instacsxsjiijcsllBSdessd,eval=FALSE,echo=toMessage}

allTxSeq <- extractTranscriptSeqs(BSgenome.Mmusculus.UCSC.mm10,
                      TxDb.Mmusculus.UCSC.mm10.knownGene,
                      use.names=TRUE)
allTxSeq
writeXStringSet(allTxSeq,
                "mm10Trans.fa")


install_CondaTools(tools="salmon", env="RNAseq_analysis")

fastaTx <- "mm10Trans.fa"
indexName <- "mm10Trans"

with_CondaEnv("RNAseq_analysis",
                      system2(command="salmon",args = c("index",
                        "-i",indexName,
                        "-t",fastaTx),
                        
                        stdout = TRUE))


fq <- "~/Downloads/filtered_ENCFF070QMF.fastq.gz"
outDir <- "TReg_2_Quant"

with_CondaEnv("RNAseq_analysis",
                      system2(command="salmon",args = c("quant",
                        "-i",indexName,
                        "-o",outDir,
                        "-l A",
                        "-r",fq),
                        
                        stdout = TRUE))


```


### 6. Review Salmon scores
Read in the generated quant.sf file and plot log2 read counts by log10 TPM scores in a scatter plot.
**NOTE: If you did not run salmon yourself, the quant file can be found here: "data/Salmon/TReg_2_Quant/quant.sf"**

```{r instacsxsjiijcsllcscdBSded,echo=toMessage}
myQuant <- read.delim("data/Salmon/TReg_2_Quant/quant.sf")
ggplot(myQuant,aes(x=NumReads,y=TPM))+geom_point()+scale_x_log10()+scale_y_log10()+theme_bw()
```
