---
title: "RNAseq in Bioconductor exercises (part 4)"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_RNAseq/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---


```{r setup, include=FALSE}
library(DESeq2)
```

In todays session we will work with some of the RNAseq data of adult mouse tissues from Bing Ren's lab, Liver and Heart.  

- More information on liver data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CHB/)
- More information on heart data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)

- Differential expression results for Heart minus Liver can be found in the data directory:  

* DGE  - **data/Heart_minus_liver.csv**
* Counts - **data/gC_TissueFull.RData**

## Exercises

### 1. Over Representaion Analysis - GO terms

Identify GO term cellular component groups enriched in genes significantly upregulated in Liver with clusterProfiler using Fishers exact test. What are the top 5 terms? Make a dotplot of these terms. Also try an enrichment map plot for the top 20 groups. 

*HINT: "CC" is the ontology category for cellular component GO terms*

```{r ea,eval=T,warning=FALSE,echo=toMessage}
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)

newResDF <- read.delim("data/Heart_minus_liver.csv",sep=",")
newResDF <- newResDF[!is.na(newResDF$padj),]

UpInLiver <- newResDF$padj < 0.05 & 
             newResDF$log2FoldChange < 0
UpInLiver_ID <- newResDF$ID[UpInLiver]

go_res <- enrichGO(UpInLiver_ID, OrgDb = org.Mm.eg.db, ont = "CC")

DT::datatable(as.data.frame(go_res)[1:5,])


dotplot(go_res, showCategory = 5)

go_res <- pairwise_termsim(go_res)
emapplot(go_res, showCategory = 20, cex_label_category=0.5)
```

### 2. Over Representaion Analysis - MSigDB

Take the same genes (significantly upregulated in Liver) and check enrichment in gene sets from MSigDB Hallmarks. Check the results with a dotplot. Rerun this but subset the universe appropriately. Compare the dotplots. 

```{r}

mm_h <- msigdbr(species = "Mus musculus", category = "H")[,c("gs_name","entrez_gene")]

sig_gene_enr1 <- enricher(UpInLiver_ID, TERM2GENE = mm_h)
dotplot(sig_gene_enr1)


sig_gene_enr2 <- enricher(UpInLiver_ID, TERM2GENE = mm_h, universe = as.character(newResDF$ID))
dotplot(sig_gene_enr2)

# the main groups are largely unchanged, but you can see a swap at the last group. This is because 15 genes from the Androgen gene set are not in the universe, so the background has shifted. 
```

### 3. GSEA for KEGG enrichment

Run GSEA on our Heart vs Liver dataset, but specifically use the KEGG genesets. Show a dotplot of the clusterProfiler results. Show the top 20 terms. Show an emap of these terms. Lastly draw the classic GSEA plots for the top 3 terms. 

*HINT: The "gseKEGG" is the clusterPorfiler function for looking at KEGG terms with GSEA*

```{r f,eval=T,warning=FALSE,echo=toMessage}

geneList <- newResDF$stat
names(geneList) <-  newResDF$ID
geneList <- geneList[order(geneList, decreasing = T)]


KEGG_HeartLiver <- gseKEGG(
  geneList,
  organism = "mmu",
  keyType = "kegg", eps = 10e-20)

head(KEGG_HeartLiver)

dotplot(KEGG_HeartLiver, showCategory = 20)


KEGG_HeartLiver_pt <- pairwise_termsim(KEGG_HeartLiver)

emapplot(KEGG_HeartLiver_pt, cex_label_category=0.5)

gseaplot(KEGG_HeartLiver_pt, geneSetID = as.data.frame(KEGG_HeartLiver_pt)$ID[1], by = "runningScore", title = as.data.frame(KEGG_HeartLiver_pt)$Description[1])

gseaplot(KEGG_HeartLiver_pt, geneSetID = as.data.frame(KEGG_HeartLiver_pt)$ID[2], by = "runningScore", title = as.data.frame(KEGG_HeartLiver_pt)$Description[2])

gseaplot(KEGG_HeartLiver_pt, geneSetID = as.data.frame(KEGG_HeartLiver_pt)$ID[3], by = "runningScore", title = as.data.frame(KEGG_HeartLiver_pt)$Description[3])
```

### 3. Additional GSEA plots for KEGG enrichment

Draw a heatmap of the genes driving enrichment in the top 3 KEGG terms. Use rlog counts, scale across row using a Z-score and include the kidney data as a reference point. 

*HINT - The gene IDs are in the core_enrichment column of the clusterProfiler result*

```{r,echo=toMessage}
library(pheatmap)
library(DESeq2)

my_gene_IDs <- unlist(strsplit(KEGG_HeartLiver$core_enrichment[1],"/"))

load("data/gC_TissueFull.RData")

dds <- DESeqDataSet(geneCounts, design = ~Tissue)
dds <- DESeq(dds)

rlogTissue <- rlog(dds)


mat_for_heatmap <- assay(rlogTissue)[rownames(rlogTissue) %in% my_gene_IDs,]

annoDF <- as.data.frame(colData(dds))
annoDF <- annoDF[,1,drop=F]


pheatmap(mat_for_heatmap,
         scale="row", show_rownames = F,
         annotation_col = annoDF)


```

