---
title: "RNAseq in Bioconductor exercises (part 3)"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_RNAseq/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---


```{r setup, include=FALSE}
library(DESeq2)
```

In todays session we will work with some of the RNAseq data of adult mouse tissues from Bing Ren's lab, Liver and Heart.  

- More information on liver data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CHB/)
- More information on heart data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)
- More information on Kidney data [can be found here ](https://www.encodeproject.org/experiments/ENCSR000CGZ/)

Preprocessed counts for Tissue data for this practical session can be found in the **data** directory.
**data/tissueCounts.RData**

## Exercises

### 1. PCA
Load in the DESeq2 object for Tissues experiment, produce an rlog transformation and plot the PCA for these genes.

```{r a,cache=TRUE,eval=TRUE,message=FALSE,warning=FALSE,echo=toMessage}
library(DESeq2)
library(ggplot2)
load("data/gC_TissueFull.RData")
dds <- DESeqDataSet(geneCounts, design = ~Tissue)
dds <- DESeq(dds)

rlogTissue <- rlog(dds)
myPlot <- DESeq2::plotPCA(rlogTissue, intgroup = "Tissue")
myPlot+theme_bw()
```

### 2. LRT test
Filter to genes significant (padj < 0..01) from our LRT test between all groups and produce a heatmap.

```{r b,cache=TRUE,eval=TRUE,message=FALSE,warning=FALSE,dependson="a",echo=toMessage}
dds2 <- DESeq(dds,test="LRT",reduced = ~1)
AllChanges <- results(dds2)
rlogMatrix <- assay(rlogTissue)
sigChanges <- rownames(AllChanges)[AllChanges$padj < 0.01 & !is.na(AllChanges$padj)]
sigMat <- rlogMatrix[rownames(rlogMatrix) %in% sigChanges,]
annoDF <- as.data.frame(colData(rlogTissue)[,1,drop=FALSE])
library(pheatmap)
pheatmap(sigMat,
         scale="row",
         show_rownames = FALSE,
         annotation_col = annoDF)

```


### 3. PCA loadings
Extract the influence of genes to PC2 and produce a heatmap as above but now ranked by PC2 rotation/loadings.

```{r c,cache=TRUE,eval=TRUE,message=FALSE,warning=FALSE,dependson="b",echo=toMessage}
pcRes <- prcomp(t(rlogMatrix))
PC2_rnk <- sort(pcRes$rotation[,2],decreasing = TRUE)
PC2_mat <- sigMat[match(names(PC2_rnk),rownames(sigMat),nomatch = 0),]
pheatmap(PC2_mat,
         scale="row",
         cluster_rows=FALSE,
         show_rownames = FALSE,annotation_col = annoDF
         )

```

### 4. PCA of a GO term
Produce a PCA plot of the GO "heart development" genes ("GO:0007507"). You will need to extract all genes within this group from the org.Mm.eg.db package. 
```{r d,cache=TRUE,eval=TRUE,message=FALSE,warning=FALSE,dependson="c",echo=toMessage}

library(org.Mm.eg.db)
HeartDevelopment <- select(org.Mm.eg.db, keytype = "GOALL",
                              keys = "GO:0007507", columns = "ENTREZID")
HeartDevelopment_Entrez <- unique(HeartDevelopment$ENTREZID)
DESeq2::plotPCA(rlogTissue[rownames(rlogTissue) %in% HeartDevelopment_Entrez],intgroup = "Tissue") + theme_bw()
```

### 5. Heatmap of a GO term
Create a Heatmap of all genes and include row annotation showing members of heart development, liver development and kidney development GO set.

```{r e,cache=TRUE,eval=TRUE,message=FALSE,warning=FALSE,dependson="d",echo=toMessage}

library(org.Mm.eg.db)
HeartDevelopment <- select(org.Mm.eg.db,keytype = "GOALL",
                              keys = "GO:0007507",columns = "ENTREZID")
HeartDevelopment_Entrez <- unique(HeartDevelopment$ENTREZID)
library(org.Mm.eg.db)
LiverDevelopment <- select(org.Mm.eg.db,keytype = "GOALL",
                              keys = "GO:0001889",columns = "ENTREZID")
LiverDevelopment_Entrez <- unique(LiverDevelopment$ENTREZID)
library(org.Mm.eg.db)
KidneyDevelopment <- select(org.Mm.eg.db,keytype = "GOALL",
                              keys = "GO:0001822",columns = "ENTREZID")
KidneyDevelopment_Entrez <- unique(KidneyDevelopment$ENTREZID)
annoRow <- data.frame(HeartDev=factor(rownames(sigMat) %in% HeartDevelopment_Entrez),
           LiverDev=factor(rownames(sigMat) %in% LiverDevelopment_Entrez),
           KidneyDev=factor(rownames(sigMat) %in% KidneyDevelopment_Entrez))
rownames(annoRow) <- rownames(sigMat)

ann_colors = list(
    HeartDev = c("FALSE"="white","TRUE"="green"),
    LiverDev = c("FALSE"="white","TRUE"="red"),
    KidneyDev = c("FALSE"="white","TRUE"="blue")
)

pheatmap(sigMat,
         scale="row",
         show_rownames = FALSE,
         annotation_col = annoDF,
         annotation_row = annoRow,
         annotation_colors = ann_colors)

```


### 6. Clustering
Take the LRT filtered rlog counts and cluster using pheatmap into 3 clusters,
```{R}
set.seed(153)
k <-   pheatmap(sigMat,
           scale="row",kmeans_k = 3)

```

### 7. Heatmap ordered by cluster

```{r km3,eval=TRUE,warning=FALSE,message=FALSE}
names(k$kmeans)
clusterDF <- as.data.frame(factor(k$kmeans$cluster))
colnames(clusterDF) <- "Cluster"
clusterDF[1:10,,drop=FALSE]

OrderByCluster <- sigMat[order(clusterDF$Cluster),]

pheatmap(OrderByCluster,
           scale="row",annotation_row = clusterDF,
           show_rownames = FALSE,cluster_rows = FALSE)

```




<!-- ### 6. Genes of Interest [ADVANCED] -->
<!-- Import the replicated peaks for H3k27ac and H3K4me3 for Heart tissue samples found in the data directory. Identify genes with both H3k4me3 and H3k27ac peaks in their TSS (+/-500bp). -->

<!-- - H3k4me3 = **data/ENCFF599BFW.bed.gz** -->
<!-- - H3k27ac = **data/ENCFF733HUI.bed.gz** -->

<!-- ```{r f,cache=TRUE,eval=TRUE,message=FALSE,warning=FALSE,dependson="e",echo=toMessage} -->
<!-- library(rtracklayer) -->
<!-- heart_H3k4me3 <- read.delim("data/ENCFF599BFW.bed.gz",sep="\t",header = FALSE) -->
<!-- heart_H3k4me3GR <- ChIPQC:::GetGRanges(heart_H3k4me3) -->

<!-- heart_H3k27ac <- read.delim("data/ENCFF733HUI.bed.gz",sep="\t",header = FALSE) -->
<!-- heart_H3k27acGR <- ChIPQC:::GetGRanges(heart_H3k27ac) -->

<!-- library(TxDb.Mmusculus.UCSC.mm10.knownGene) -->
<!-- genePos <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene) -->
<!-- TSSPos <- promoters(genePos,500,500) -->

<!-- ActiveGenes <- unique(TSSPos$gene_id[TSSPos %over% heart_H3k4me3GR  -->
<!--                          & TSSPos %over% heart_H3k27acGR]) -->

<!-- ``` -->


<!-- Create a heatmap of the top 100 marker (both positive and negative) for PC1 and add annotation for genes with H3k27ac and H3K4me3 peaks in their TSS. [ADVANCED] -->

<!-- ```{r g,cache=TRUE,eval=TRUE,message=FALSE,warning=FALSE,dependson="f",echo=toMessage} -->
<!-- library(pheatmap) -->
<!-- annoRow <- data.frame(ActiveHeart=(rownames(sigMat) %in% ActiveGenes)+0, -->
<!--                       row.names = rownames(sigMat)) -->


<!-- PC1markers <- sort(pcRes$rotation[,1],decreasing = FALSE)[1:100] -->
<!-- sigMats <- rlogMatrix[rownames(rlogMatrix) %in% names(PC1markers),] -->
<!-- pheatmap(sigMats, -->
<!--          scale="row",annotation_row = annoRow, -->
<!--          show_rownames = FALSE) -->

<!-- PC1markers <- sort(pcRes$rotation[,1],decreasing = TRUE)[1:100] -->
<!-- sigMats <- rlogMatrix[rownames(rlogMatrix) %in% names(PC1markers),] -->
<!-- pheatmap(sigMats, -->
<!--          scale="row",annotation_row = annoRow, -->
<!--          show_rownames = FALSE) -->

<!-- ``` -->



<!-- ```{r ea,cache=TRUE,eval=F,dependson="d",warning=FALSE,echo=toMessage} -->
<!-- newResDF <- read.delim("data/Heart_minus_liver.csv",sep=",") -->
<!-- newResDF <- newResDF[!is.na(newResDF$padj),] -->

<!-- UpInLiver <- newResDF$padj < 0.05 &  -->
<!--              newResDF$log2FoldChange < 0 -->
<!-- UpInLiver <- as.integer(UpInLiver) -->
<!-- names(UpInLiver) <- newResDF$ID -->
<!-- library(goseq) -->
<!-- pwf = nullp(UpInLiver, "mm10", "knownGene", plot.fit = TRUE) -->
<!-- Kegg_UpInLiver <- goseq(pwf,"mm10","knownGene", -->
<!--                        test.cats=c("KEGG")) -->
<!-- library(KEGG.db) -->
<!-- xx <- as.list(KEGGPATHID2NAME) -->
<!-- idtoName <- cbind(names(xx),unlist(xx)) -->

<!-- Kegg_UpInLiver <- merge(idtoName,Kegg_UpInLiver,by=1,all=TRUE) -->
<!-- orderByP <- order(Kegg_UpInLiver$over_represented_pvalue) -->
<!-- Kegg_UpInLiver <- Kegg_UpInLiver[orderByP,] -->
<!-- ``` -->

