<!DOCTYPE html>
<html>
  <head>
    <title></title>
    <meta charset="utf-8">
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Analysis of RNA-seq data in R and Bioconductor
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### <a href="http://rockefelleruniversity.github.io/RU_RNAseq/" class="uri">http://rockefelleruniversity.github.io/RU_RNAseq/</a>

---





                                 
## The data.

In this session we will use one of our datasets from a splicing factor Knockdown RNAseq experiment found at GEO [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE101763).

I have aligned all FQ to BAM and counted in genes and exons using Rsubread and summariseOverlaps().



---
## The data.

All Gene and Exon counts can be found as .RData obects in **Data** directory

- Counts in disjoint exons can be found at - **Data/senescence_ExonCounts.RData**

---
## What we will cover.

In our first session we will looked at two separate ways we can gain gene expression estimates from raw sequencing data as fastQ files.

In our last session we imported our expression estimates to identify differences in the expression of genes between two conditions. 

In this session we will look at how we can identify differences in the expression of genes between multiple conditions and how we can identify changes in Exon usage to infer changes in transcript usage.

---
## Counts from SummariseOverlaps

In our first session we counted our reads in genes using the summarise overlaps function to generate our **RangedSummarizedExperiment** object.

I have counted using a **BamFileList** of all T-Reg samples to easily organise my samples and save memory with the **yieldSize** parameter.


```r
library(Rsamtools)
bamFilesToCount &lt;- c("Sorted_shPTBP1_53_rep1.bam","Sorted_shPTBP1_53_rep2.bam","Sorted_shPTBP1_53_rep3.bam",
                     "Sorted_vec_4OHT_rep1.bam","Sorted_vec_4OHT_rep2.bam","Sorted_vec_4OHT_rep3.bam" )
myBams &lt;- BamFileList(bamFilesToCount,yieldSize = 10000)
```

---
## Counts from SummariseOverlaps

For differential exon usage we must use a non-overlapping,disjoint set of exons to count on.

We can produce a non-overlapping set of exons by using the **disjointExons()** function. This is equivalent of us having to create GFF file and using the python script wrapped up in DEXseq package (*dexseq_prepare_annotation.py*).


```r
library(GenomicFeatures)
nonOverlappingExons &lt;- disjointExons(TxDb.Hsapiens.UCSC.hg19.knownGene)
names(nonOverlappingExons) &lt;- paste(mcols(nonOverlappingExons)$gene_id,
                                    mcols(nonOverlappingExons)$exonic_part,
                                    sep="_")
nonOverlappingExons[1:2,]
```

```
## GRanges object with 2 ranges and 3 metadata columns:
##       seqnames               ranges strand |         gene_id
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;
##   1_1    chr19 [58858172, 58858395]      - |               1
##   1_2    chr19 [58858719, 58859006]      - |               1
##               tx_name exonic_part
##       &lt;CharacterList&gt;   &lt;integer&gt;
##   1_1      uc002qsd.4           1
##   1_2      uc002qsd.4           2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Disjoint Exons

![](imgs/collapsedE1.png)

---
## Disjoint Exons

![](imgs/collapsedE2.png)


---
## Counts from SummariseOverlaps

Now we have our disjoint exons to count on we can use the **summarizeOverlaps().**



```r
senescence_ExonCounts &lt;- summarizeOverlaps(nonOverlappingExons,
                                    myBam,
                                    ignore.strand = TRUE,
                                    inter.feature=FALSE)
senescence_ExonCounts
```


```
## class: RangedSummarizedExperiment 
## dim: 269121 6 
## metadata(0):
## assays(1): counts
## rownames(269121): 1_1 1_2 ... 9997_3 9997_4
## rowData names(3): gene_id tx_name exonic_part
## colnames(6): Sorted_shPTBP1_53_rep1.bam Sorted_shPTBP1_53_rep2.bam
##   ... Sorted_vec_4OHT_rep2.bam Sorted_vec_4OHT_rep3.bam
## colData names(1): condition
```

---
## Counts from SummariseOverlaps

Our **RangedSummarizedExperiment** contains our counts as a matrix accessible by the **assay()** function.

Each row is named by our Exon ID, geneID + exon number.



```r
assay(senescence_ExonCounts)[1:2,]
```

```
##     Sorted_shPTBP1_53_rep1.bam Sorted_shPTBP1_53_rep2.bam
## 1_1                        400                        489
## 1_2                        573                        970
##     Sorted_shPTBP1_53_rep3.bam Sorted_vec_4OHT_rep1.bam
## 1_1                        317                      326
## 1_2                        455                      452
##     Sorted_vec_4OHT_rep2.bam Sorted_vec_4OHT_rep3.bam
## 1_1                      650                      508
## 1_2                      801                      885
```
---
## Counts from SummariseOverlaps

We can retrieve the GRanges we counted by the **rowRanges()** accessor function.

The GRanges rows contains our disjoint exons used in counting and is in matching order as rows in our count table.



```r
rowRanges(senescence_ExonCounts)[1:5,]
```

```
## GRanges object with 5 ranges and 3 metadata columns:
##       seqnames               ranges strand |         gene_id
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;
##   1_1    chr19 [58858172, 58858395]      - |               1
##   1_2    chr19 [58858719, 58859006]      - |               1
##   1_3    chr19 [58859832, 58860494]      - |               1
##   1_4    chr19 [58860934, 58861735]      - |               1
##   1_5    chr19 [58861736, 58862017]      - |               1
##                     tx_name exonic_part
##             &lt;CharacterList&gt;   &lt;integer&gt;
##   1_1            uc002qsd.4           1
##   1_2            uc002qsd.4           2
##   1_3            uc002qsf.2           3
##   1_4            uc002qsf.2           4
##   1_5 uc002qsd.4,uc002qsf.2           5
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```


---
## Differential transcript usage from RNAseq data.

Previously we have investigated how we may identify changes in the overall abundance on gene.

![](imgs/dge.png)

---
## Differential transcript usage from RNAseq data.

We may also be interested in identifying any evidence of a changes in transcript usage within a gene. 

![](imgs/dtu.png)

---
## Differential transcript usage from RNAseq data.

Due to the uncertainty in the assignment of reads to overlapping transcripts from the same genes, a common approach to investigate differential transcript usage is to define non-overlapping disjoint exons and identify changes in their usage. 


![](imgs/deu2.png)


---
## DEXSeq

Options for differential transcript usage are available in R including DEXseq, and Voom/Limma.

DEXseq can be a little slow when fitting models but it has options for parallization. 

Voom/Limma is significantly faster by comparison but provides a less informative output. 

---
## DEXSeq

To use DEXseq package has a similar set up to that seen for DESeq2.

We first must define our metadata data.frame of sample groups, with sample rownames matching colnames of our **RangedSummarizedExperiment** object. 


```r
metaData &lt;- data.frame(condition=c("shPTBP1_53", "shPTBP1_53", "shPTBP1_53","senescence","senescence","senescence"),
                       row.names = colnames(senescence_ExonCounts))
metaData
```

```
##                             condition
## Sorted_shPTBP1_53_rep1.bam shPTBP1_53
## Sorted_shPTBP1_53_rep2.bam shPTBP1_53
## Sorted_shPTBP1_53_rep3.bam shPTBP1_53
## Sorted_vec_4OHT_rep1.bam   senescence
## Sorted_vec_4OHT_rep2.bam   senescence
## Sorted_vec_4OHT_rep3.bam   senescence
```


---
## DEXseq

Previously we have been using **DESeqDataSetFromMatrix()** constructor for DESeq2. In DEXSeq we can use the **DEXSeqDataSet** constructor in a similar manner.

As with the **DESeqDataSetFromMatrix()** function, we must provide the counts matrix, our metadata dataframe and our design specifying with metadata columns to test.

In addition to this we must provide our exon IDs and gene IDs to the allow for the association of exons to the genes they constitute.


```r
countMatrix &lt;- assay(senescence_ExonCounts)
countGRanges &lt;- rowRanges(senescence_ExonCounts)
geneIDs &lt;- as.vector(unlist(countGRanges$gene_id))
exonIDs &lt;- rownames(countMatrix)
```



---
## DEXseq


We must specify our design to include a sample and exon effect to distinguish change in exon expression from a change in exon usage.

Previously we simply added a design formula with our group of interest.

```
~ GROUPOFINTERERST
```

Now we specify.

```
~ sample + exon + GROUPOFINTERERST:exon
```

---
## DEXseq

Now we can use the **DEXSeqDataSet** constructor as we have **DESeqDataSetFromMatrix()** but in addition we specify the exon ids to the **featureID** argument and the gene ids to the **groupID** argument.



```r
ddx &lt;- DEXSeqDataSet(countMatrix,metaData,
                     design= ~ sample + exon + condition:exon,
                     featureID = exonIDs,
                     groupID = geneIDs)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 269121 12 
## metadata(1): version
## assays(1): counts
## rownames(269121): 1:1_1 1:1_2 ... 9997:9997_3 9997:9997_4
## rowData names(4): featureID groupID exonBaseMean exonBaseVar
## colnames: NULL
## colData names(3): sample condition exon
```

---
## DEXseq from  RangedSummarizedExperiment

As we have generated a **RangedSummarizedExperiment** for exon counting, we can use the more straightforward  **DEXSeqDataSetFromSE** function as we used the **DESeqDataSet** function.

We can simply provide our **RangedSummarizedExperiment** and design to the **DEXSeqDataSetFromSE** function.



```r
colData(senescence_ExonCounts)$condition &lt;- c("shPTBP1_53", "shPTBP1_53", "shPTBP1_53",
                                       "senescence","senescence","senescence") 
ddx &lt;- DEXSeqDataSetFromSE(senescence_ExonCounts,
                     design= ~ sample + exon + condition:exon)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 269121 12 
## metadata(1): version
## assays(1): counts
## rownames(269121): 1:E001 1:E002 ... 9997:E003 9997:E004
## rowData names(5): featureID groupID exonBaseMean exonBaseVar
##   transcripts
## colnames: NULL
## colData names(3): sample condition exon
```

---
## DEXseq (filtering low exon counts)

DEXseq is fairly computationally expensive so we can reduce the time taken by removing exons with little expression up front.

Here we can filter any exons which do not have 10 reads in at least two samples (our smallest group size is 3).


```r
ToFilter &lt;- apply(assay(senescence_ExonCounts),1,function(x)sum(x &gt; 10)) &gt; 2
senescence_ExonCounts &lt;- senescence_ExonCounts[ToFilter,]
table(ToFilter)
```

```
## ToFilter
##  FALSE   TRUE 
## 113251 155870
```

```r
ddx &lt;- DEXSeqDataSetFromSE(senescence_ExonCounts,
                     design= ~ sample + exon + condition:exon)
ddx
```

```
## class: DEXSeqDataSet 
## dim: 155870 12 
## metadata(1): version
## assays(1): counts
## rownames(155870): 1:E001 1:E002 ... 9997:E001 9997:E003
## rowData names(5): featureID groupID exonBaseMean exonBaseVar
##   transcripts
## colnames: NULL
## colData names(3): sample condition exon
```

---
## Running DEXseq in steps.

We can normalize, estimate dispersion and test for differential exon usage with the **estimateSizeFactors()**, **estimateDispersions()** and **testForDEU()** functions. 

Finally we can estimate fold changes for exons using the **estimateExonFoldChanges()** function and specifying metadata column of interest to **fitExpToVar** parameter.


```r
dxd &lt;- estimateSizeFactors(dxd)
dxd &lt;- estimateDispersions(dxd)
dxd &lt;- testForDEU(dxd, reducedModel = ~ sample + exon)
dxd = estimateExonFoldChanges( dxd, fitExpToVar="condition")
```


---
## DEXseq results.

Once have processed our data through the **DEXseq** function we can use the **DEXSeqResults()** function with our **DEXseq** object. We order table to arrange by pvalue.




```r
dxr1 &lt;- DEXSeqResults(dxd)
dxr1 &lt;- dxr1[order(dxr1$pvalue),]
```




---
## DEXseq function.

As with DESeq2, the DEXseq has a single workflow function which will normalise to library sizes, estimate dispersion and perform differential exon usage in one go. The returned object is our **DEXSeqResults** results object.


```r
dxr1 &lt;- DEXSeq(ddx)
dxr1 &lt;- dxr1[order(dxr1$pvalue),]
```

---
## DEXseq results.

The **DEXSeqResults** results object contains significantly more information than the DESeq2 results objects we have seen so far.


```r
dxr1[1,]
```

```
## 
## LRT p-value: full vs reduced
## 
## DataFrame with 1 row and 13 columns
##                groupID   featureID exonBaseMean dispersion      stat
##            &lt;character&gt; &lt;character&gt;    &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
## 57142:E008       57142        E008     21482.14 0.00455953  2036.225
##               pvalue      padj senescence shPTBP1_53
##            &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;
## 57142:E008         0         0   30.18552   63.80855
##            log2fold_shPTBP1_53_senescence              genomicData
##                                 &lt;numeric&gt;                &lt;GRanges&gt;
## 57142:E008                       4.121217 chr2:55252222-55254621:-
##                        countData           transcripts
##                         &lt;matrix&gt;                &lt;list&gt;
## 57142:E008 40618:35656:54755:... uc002ryd.3,uc002rye.3
```

---
## DEXseq results.

The **DEXSeqResults** results object can be converted to a data.frame as with **DESeqResults** objects


```r
as.data.frame(dxr1)[1,]
```

```
##            groupID featureID exonBaseMean dispersion     stat pvalue padj
## 57142:E008   57142      E008     21482.14 0.00455953 2036.225      0    0
##            senescence shPTBP1_53 log2fold_shPTBP1_53_senescence
## 57142:E008   30.18552   63.80855                       4.121217
##            genomicData.seqnames genomicData.start genomicData.end
## 57142:E008                 chr2          55252222        55254621
##            genomicData.width genomicData.strand
## 57142:E008              2400                  -
##            countData.Sorted_shPTBP1_53_rep1.bam
## 57142:E008                                40618
##            countData.Sorted_shPTBP1_53_rep2.bam
## 57142:E008                                35656
##            countData.Sorted_shPTBP1_53_rep3.bam
## 57142:E008                                54755
##            countData.Sorted_vec_4OHT_rep1.bam
## 57142:E008                               1007
##            countData.Sorted_vec_4OHT_rep2.bam
## 57142:E008                                747
##            countData.Sorted_vec_4OHT_rep3.bam  transcripts
## 57142:E008                                838 uc002ryd....
```

---
## DEXseq results.

A very useful feature of the **DEXSeqResults** results objects is the ability to visualise the differential exons along our gene using the **plotDEXSeq()** function and providing our **DEXSeqResults** results object and gene of interest to plot.


```r
plotDEXSeq(dxr1,"57142")
```

![](RU_RNAseq_p_3.5_files/figure-html/da4-1.png)&lt;!-- --&gt;

---
## DEXseq results.

We can specify the parameter **displayTranscripts** to TRUE to see the full gene model.

```r
plotDEXSeq(dxr1,"57142",displayTranscripts=FALSE)
```

![](RU_RNAseq_p_3.5_files/figure-html/da5c-1.png)&lt;!-- --&gt;

---
## DEXseq results.

![](imgs/RTN4.png)



---
## DEXseq results.

The differential exon usage for Entrez gene 7168 is more complex.


```r
plotDEXSeq(dxr1,"7168",displayTranscripts=TRUE)
```

![](RU_RNAseq_p_3.5_files/figure-html/da5-1.png)&lt;!-- --&gt;

---
## DEXseq results.


![](imgs/TPM1.png)

---
## DEXseq results.

We can produce an HTML report for selected genes using the **DEXSeqHTML** function. By default this will create a file called *testForDEU.html* within a folder *DEXSeqReport* in the current working directory.


```r
DEXSeqHTML(dxr1,"57142")
```

[Link to result here](../../Data/DEXSeqReport/testForDEU.html)

---
## DEXseq results- Gene Q values.

It may be desirable to obtain a pvalue summarising the differential exon usage to a gene level.

We can obtain an adjusted p-value per gene with the **perGeneQValue()** with our **DEXSeqResults** results object.

The result is a vector of adjusted p-values named by their corresponding gene.


```r
geneQ &lt;- perGeneQValue(dxr1)
geneQ[1:10]
```

```
##         1       100      1000     10000     10001 100033414 100037417 
##         1         1         1         1         1         1         1 
## 100048912 100049076     10005 
##         1         1         1
```

---
## DEXseq results- Gene Q values.

We could merge this back to our full results table to provides exon and gene level statistics in one report.


```r
gQFrame &lt;- as.data.frame(geneQ,row.names = names(geneQ))
dxDF &lt;- as.data.frame(dxr1)
dxDF &lt;-merge(gQFrame,dxDF,by.x=0,by.y=1,all=TRUE)
dxDF &lt;- dxDF[order(dxDF$geneQ,dxDF$pvalue),]
dxDF[1:2,]
```

```
##        Row.names geneQ featureID exonBaseMean  dispersion     stat
## 97377      57142     0      E008    21482.144 0.004559530 2036.225
## 117732      7168     0      E009     2805.899 0.004776165 1092.222
##               pvalue          padj senescence shPTBP1_53
## 97377   0.000000e+00  0.000000e+00   30.18552   63.80855
## 117732 1.619401e-239 1.246016e-234   49.34291   18.36318
##        log2fold_shPTBP1_53_senescence genomicData.seqnames
## 97377                        4.121217                 chr2
## 117732                      -4.264727                chr15
##        genomicData.start genomicData.end genomicData.width
## 97377           55252222        55254621              2400
## 117732          63353397        63353472                76
##        genomicData.strand countData.Sorted_shPTBP1_53_rep1.bam
## 97377                   -                                40618
## 117732                  +                                  619
##        countData.Sorted_shPTBP1_53_rep2.bam
## 97377                                 35656
## 117732                                  668
##        countData.Sorted_shPTBP1_53_rep3.bam
## 97377                                 54755
## 117732                                  705
##        countData.Sorted_vec_4OHT_rep1.bam
## 97377                                1007
## 117732                               4442
##        countData.Sorted_vec_4OHT_rep2.bam
## 97377                                 747
## 117732                               4979
##        countData.Sorted_vec_4OHT_rep3.bam  transcripts
## 97377                                 838 uc002ryd....
## 117732                               5206 uc002alh....
```


---
## RNAseq with multiple groups.

So far we have only reviewed a two group comparison of differential expression and differential transcript usage.

We can compare multiple groups fairly easily in DESeq2.

---
## RNAseq with multiple groups.

For this we will be using the Tissue RNAseq datasets from Encode. 

I have pre-aligned and counted the Tissue RNAseq datasets and provide a **RangedSummarizedExperiment** object

**../Data/gC_TissueFull.RData**

---
## RNAseq with multiple groups.

I load our precounted Tissues' reads gene model and the loaded **RangedSummarizedExperiment** contains counts from three groups.


```r
load("../../Data/gC_TissueFull.RData")
geneCounts_Tissue
```

```
## class: RangedSummarizedExperiment 
## dim: 24100 6 
## metadata(0):
## assays(1): ''
## rownames(24100): 497097 19888 ... 100040911 170942
## rowData names(0):
## colnames(6): Sorted_Heart_1 Sorted_Heart_2 ... Sorted_Liver_1
##   Sorted_Liver_2
## colData names(0):
```

---
## RNAseq with multiple groups.

We can update the **colData** slot of our **RangedSummarizedExperiment** object to allow us to make use of metadata in DESeq2 later.


```r
colData(geneCounts_Tissue)$Tissue &lt;- c("Heart","Heart",
                                       "Kidney","Kidney",
                                       "Liver","Liver")
```

---
## RNAseq with multiple groups.

We can now set up our **DESeq2** object as we have previously and run the **DESeq** workflow function


```r
dds &lt;- DESeqDataSet(geneCounts_Tissue,design = ~Tissue)
dds &lt;- DESeq(dds)
```

---
## RNAseq with multiple groups.

To extract comparisons now we can simply specify the tissues of interest to the results funtion.


```r
heartVsLiver &lt;- results(dds,c("Tissue","Heart","Liver"))
heartVskidney &lt;- results(dds,c("Tissue","Heart","Kidney"))
heartVskidney
```

```
## log2 fold change (MLE): Tissue Heart vs Kidney 
## Wald test p-value: Tissue Heart vs Kidney 
## DataFrame with 24100 rows and 6 columns
##              baseMean log2FoldChange     lfcSE        stat       pvalue
##             &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;    &lt;numeric&gt;
## 497097       1.709418     -0.1114035 1.9683195 -0.05659826 9.548652e-01
## 19888        6.073255      1.6401703 0.9332174  1.75754359 7.882520e-02
## 20671       64.374826      1.6542972 0.4130024  4.00553853 6.187635e-05
## 27395      352.706140      1.0491202 0.2780073  3.77371523 1.608343e-04
## 18777      517.655886     -1.9976114 0.3406895 -5.86343629 4.533849e-09
## ...               ...            ...       ...         ...          ...
## 434935     0.00000000             NA        NA          NA           NA
## 100041256  0.00000000             NA        NA          NA           NA
## 100039574  0.00000000             NA        NA          NA           NA
## 100040911  0.09745882     -0.4459651 4.9965270 -0.08925502    0.9288792
## 170942    10.24190603     -1.3308086 0.9173729 -1.45067346    0.1468708
##                   padj
##              &lt;numeric&gt;
## 497097    9.786147e-01
## 19888     1.666689e-01
## 20671     2.804627e-04
## 27395     6.811435e-04
## 18777     3.523962e-08
## ...                ...
## 434935              NA
## 100041256           NA
## 100039574           NA
## 100040911           NA
## 170942       0.2734518
```

---
## RNAseq with multiple groups.

To identify genes specifically upregulated in Heart versus other tissues. We can then overlap the upregulated genes from the Heart vs Liver and  Heart vs Kidney comparisons.

First we convert **DESeqResults** objects  into data.frame, remove NAs and extract most interesting columns to us.


```r
heartVsLiverDF &lt;- as.data.frame(heartVsLiver)
heartVskidneyDF &lt;- as.data.frame(heartVskidney)
heartVsLiverDF &lt;- heartVsLiverDF[!is.na(heartVsLiverDF$padj),]
heartVskidneyDF &lt;- heartVskidneyDF[!is.na(heartVskidneyDF$padj),]
heartVsLiverDF &lt;- heartVsLiverDF[,c("log2FoldChange","padj")]
heartVskidneyDF &lt;- heartVskidneyDF[,c("log2FoldChange","padj")]
```

---
## RNAseq with multiple groups.

We can then update the column names and merge our data.frame to have a single table of most useful information.


```r
colnames(heartVskidneyDF) &lt;- paste0("HeartVsKidney","_",colnames(heartVskidneyDF))
colnames(heartVsLiverDF) &lt;- paste0("HeartVsLiver","_",colnames(heartVsLiverDF))
fullTable &lt;- merge(heartVsLiverDF,heartVskidneyDF,by=0)
fullTable[1:2,]
```

```
##   Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 1    100017                  -0.5992113         0.2118830
## 2    100019                  -0.6497866         0.1418237
##   HeartVsKidney_log2FoldChange HeartVsKidney_padj
## 1                    0.1578808          0.8033665
## 2                   -0.1634218          0.7817889
```

---
## RNAseq with multiple groups.

Now we can extract our genes upregulated in Heart in both conditions.


```r
upInHeart &lt;- fullTable$HeartVsLiver_log2FoldChange &gt; 0 &amp;
             fullTable$HeartVsKidney_log2FoldChange &gt; 0 &amp;
             fullTable$HeartVsLiver_padj &lt; 0.05 &amp;
             fullTable$HeartVsKidney_padj &lt; 0.05
upInHeartTable &lt;- fullTable[upInHeart,]
upInHeartTable[1:2,]
```

```
##    Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 15 100038347                    4.950045      1.057860e-49
## 37 100038575                    2.454009      2.156496e-02
##    HeartVsKidney_log2FoldChange HeartVsKidney_padj
## 15                     4.342545       2.003331e-50
## 37                     3.161309       1.038031e-03
```

---
## RNAseq with multiple groups.

We can also make a logical data.frame of whether a gene was upregulated in Heart for both Liver and Kidney comparisons.


```r
forVenn &lt;- data.frame(UpvsLiver=fullTable$HeartVsLiver_log2FoldChange &gt; 0 &amp;
                       fullTable$HeartVsLiver_padj &lt; 0.05,
                     UpvsKidney=fullTable$HeartVsKidney_log2FoldChange &gt; 0 &amp;
                       fullTable$HeartVsKidney_padj &lt; 0.05)
forVenn[1:3,]
```

```
##   UpvsLiver UpvsKidney
## 1     FALSE      FALSE
## 2     FALSE      FALSE
## 3     FALSE       TRUE
```

---
## RNAseq with multiple groups.

We can use **limmas'** **vennDiagram()** to produce a venn diagram of overlap in upregulation in heart vs Liver or Kidney. 


```r
library(limma)
vennDiagram(forVenn)
```

![](RU_RNAseq_p_3.5_files/figure-html/tissue10-1.png)&lt;!-- --&gt;

---
## RNAseq with multiple groups.

When dealing with multiple groups it may be of interest to identify genes are highly variable between groups but with low variability within groups, akin to an Anova test.

In DESeq2, we can identify genes significantly changing across groups by comparing our models with and without our groups of interest using the **results** function.

- Tissue model

```
~ Tissue
```

- No groups model

```
~ 1
```

---
## RNAseq with multiple groups.

To compare out two models we must set the parameter of **reduced** to our alternative model of no groups. We also set the **test** parameter to **LRT** to allow us to compare the models.


```r
dds2 &lt;- DESeq(dds,test="LRT",reduced=~1)
acrossGroups &lt;- results(dds2)
acrossGroups &lt;- acrossGroups[order(acrossGroups$pvalue),]
acrossGroups[1:3,]
```

```
## log2 fold change (MLE): Tissue Liver vs Heart 
## LRT p-value: '~ Tissue' vs '~ 1' 
## DataFrame with 3 rows and 6 columns
##        baseMean log2FoldChange     lfcSE      stat        pvalue
##       &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;
## 16483  125680.4      -2.294864 1.5830283  1642.173  0.000000e+00
## 17888   43583.9     -14.874592 0.6524674  1566.142  0.000000e+00
## 21956   16505.0     -15.058827 1.0578142  1430.916 1.908172e-311
##                padj
##           &lt;numeric&gt;
## 16483  0.000000e+00
## 17888  0.000000e+00
## 21956 1.095545e-307
```


---
## RNAseq with multiple groups.

We can use the **plotCounts()** to get review expression profile of a gene, one at a time. We define the gene of interest to **gene** parameter and **intgroup** to specify metadata column to group counts by.


```r
plotCounts(dds2,gene="17888",intgroup = "Tissue")
```

![](RU_RNAseq_p_3.5_files/figure-html/tissue12-1.png)&lt;!-- --&gt;

---
## RNAseq with multiple groups.

We can combine our result with our previous table of pair-wise contrasts to provide information on where differences between groups exist.


```r
acrossGroups &lt;- as.data.frame(acrossGroups[,c("pvalue","padj")])
colnames(acrossGroups) &lt;- paste0("AllGroups","_",colnames(acrossGroups))
allGroups &lt;- merge(fullTable,acrossGroups,by.x=1,by.y=0,all=TRUE)
allGroups &lt;- allGroups[order(allGroups$AllGroups_pvalue),]
allGroups[1:2,]
```

```
##      Row.names HeartVsLiver_log2FoldChange HeartVsLiver_padj
## 5257     16483                    2.294864      2.469609e-01
## 6057     17888                   14.874592     1.633323e-111
##      HeartVsKidney_log2FoldChange HeartVsKidney_padj AllGroups_pvalue
## 5257                    -16.69262      2.113510e-140                0
## 6057                     14.84458      2.139909e-218                0
##      AllGroups_padj
## 5257              0
## 6057              0
```


---

# Time for an exercise.

[Link_to_exercises](../../Exercises/RNAseq_part3_exercises.html)

[Link_to_answers](../../Answers/RNAseq_part3_answers.html)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
